// @po-ui/ng-components/esm2015/lib/components/po-stepper/po-stepper-circle/po-stepper-circle.component.js
import { __decorate, __metadata } from "tslib";
import { Component, Input } from '@angular/core';
import { PoStepperStatus } from '../enums/po-stepper-status.enum';
const poMediumStepSize = 32;
const poLargeStepSize = 48;
/**
 * @docsPrivate
 *
 * @description
 *
 * Componente responsável por exibir os círculos dos *steps*.
 */
let PoStepperCircleComponent = class PoStepperCircleComponent {
    get isActive() {
        return this.status === PoStepperStatus.Active;
    }
    get isDefault() {
        return this.status === PoStepperStatus.Default;
    }
    get isDisabled() {
        return this.status === PoStepperStatus.Disabled;
    }
    get isDone() {
        return this.status === PoStepperStatus.Done;
    }
    get isError() {
        return this.status === PoStepperStatus.Error;
    }
    get isLargeStep() {
        return this.size >= poLargeStepSize;
    }
    get isMediumStep() {
        return this.size >= poMediumStepSize && !this.isLargeStep;
    }
};
__decorate([
    Input('p-content'),
    __metadata("design:type", Object)
], PoStepperCircleComponent.prototype, "content", void 0);
__decorate([
    Input('p-icons'),
    __metadata("design:type", Boolean)
], PoStepperCircleComponent.prototype, "icons", void 0);
__decorate([
    Input('p-size'),
    __metadata("design:type", Number)
], PoStepperCircleComponent.prototype, "size", void 0);
__decorate([
    Input('p-status'),
    __metadata("design:type", String)
], PoStepperCircleComponent.prototype, "status", void 0);
PoStepperCircleComponent = __decorate([
    Component({
        selector: 'po-stepper-circle',
        template: "<div\n  class=\"po-stepper-circle\"\n  [class.po-stepper-circle-with-icon]=\"icons\"\n  [style.height.px]=\"size\"\n  [style.width.px]=\"size\"\n  [tabindex]=\"isDisabled ? -1 : 0\"\n>\n  <span\n    class=\"po-stepper-circle-content\"\n    [class.po-icon]=\"icons\"\n    [class.po-icon-exclamation]=\"icons && isError\"\n    [class.po-icon-info]=\"icons && (isActive || isDefault || isDisabled)\"\n    [class.po-icon-ok]=\"icons && isDone\"\n    [class.po-stepper-circle-content-lg]=\"isLargeStep\"\n    [class.po-stepper-circle-content-md]=\"isMediumStep\"\n  >\n    {{ !icons ? content : '' }}\n  </span>\n</div>\n"
    })
], PoStepperCircleComponent);
export { PoStepperCircleComponent };
