// @po-ui/ng-components/esm2015/lib/components/po-table/po-table-subtitle-circle/po-table-subtitle-circle.component.js
import { __decorate, __metadata } from "tslib";
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
import { PoColorPaletteService } from './../../../services/po-color-palette/po-color-palette.service';
/**
 * @docsPrivate
 *
 * @description
 *
 * Componente para a criação da representação da legenda, em formato de círculo.
 */
let PoTableSubtitleCircleComponent = class PoTableSubtitleCircleComponent {
    constructor(poColorPaletteService) {
        this.poColorPaletteService = poColorPaletteService;
        /** Esconde a tooltip. */
        this.hideTitle = false;
    }
    /** Objeto com os dados da legenda. */
    set subtitle(subtitle) {
        if (subtitle) {
            subtitle.color = this.poColorPaletteService.getColor(subtitle);
        }
        this._subtitle = subtitle;
    }
    get subtitle() {
        return this._subtitle;
    }
};
PoTableSubtitleCircleComponent.ctorParameters = () => [
    { type: PoColorPaletteService }
];
__decorate([
    Input('p-subtitle'),
    __metadata("design:type", Object),
    __metadata("design:paramtypes", [Object])
], PoTableSubtitleCircleComponent.prototype, "subtitle", null);
__decorate([
    Input('p-hide-title'),
    __metadata("design:type", Boolean)
], PoTableSubtitleCircleComponent.prototype, "hideTitle", void 0);
PoTableSubtitleCircleComponent = __decorate([
    Component({
        selector: 'po-table-subtitle-circle',
        template: "<span [title]=\"hideTitle ? '' : subtitle?.label\" class=\"po-table-subtitle-circle po-{{ subtitle?.color }}\">\n  {{ subtitle?.content }}\n</span>\n",
        changeDetection: ChangeDetectionStrategy.OnPush
    }),
    __metadata("design:paramtypes", [PoColorPaletteService])
], PoTableSubtitleCircleComponent);
export { PoTableSubtitleCircleComponent };
