// @po-ui/ng-components/esm2015/lib/components/po-field/po-upload/po-upload-drag-drop/po-upload-drag-drop-area-overlay/po-upload-drag-drop-area-overlay.component.js
import { __decorate, __metadata } from "tslib";
import { AfterViewInit, ChangeDetectionStrategy, Component, ElementRef, EventEmitter, Input, Output, Renderer2, ViewChild } from '@angular/core';
let PoUploadDragDropAreaOverlayComponent = class PoUploadDragDropAreaOverlayComponent {
    constructor(renderer) {
        this.renderer = renderer;
        this.areaElement = new EventEmitter();
    }
    ngAfterViewInit() {
        if (this.target) {
            this.setPosition(this.target);
        }
        this.areaElement.emit(this.DragDropAreaFixed.nativeElement);
    }
    setPosition(targetElement) {
        const boundingClientRect = targetElement.nativeElement.getBoundingClientRect();
        const clientRectKeys = ['bottom', 'left', 'height', 'right', 'top', 'width'];
        clientRectKeys.forEach(clientRectKey => {
            const clientRectValue = boundingClientRect[clientRectKey];
            this.renderer.setStyle(this.DragDropAreaFixed.nativeElement, clientRectKey, `${clientRectValue}px`);
        });
    }
};
PoUploadDragDropAreaOverlayComponent.ctorParameters = () => [
    { type: Renderer2 }
];
__decorate([
    ViewChild('DragDropAreaFixed', { read: ElementRef, static: true }),
    __metadata("design:type", ElementRef)
], PoUploadDragDropAreaOverlayComponent.prototype, "DragDropAreaFixed", void 0);
__decorate([
    Input('p-directory-compatible'),
    __metadata("design:type", Boolean)
], PoUploadDragDropAreaOverlayComponent.prototype, "directoryCompatible", void 0);
__decorate([
    Input('p-disabled'),
    __metadata("design:type", Boolean)
], PoUploadDragDropAreaOverlayComponent.prototype, "disabled", void 0);
__decorate([
    Input('p-literals'),
    __metadata("design:type", Object)
], PoUploadDragDropAreaOverlayComponent.prototype, "literals", void 0);
__decorate([
    Input('p-target'),
    __metadata("design:type", ElementRef)
], PoUploadDragDropAreaOverlayComponent.prototype, "target", void 0);
__decorate([
    Output('p-area-element'),
    __metadata("design:type", EventEmitter)
], PoUploadDragDropAreaOverlayComponent.prototype, "areaElement", void 0);
PoUploadDragDropAreaOverlayComponent = __decorate([
    Component({
        selector: 'po-upload-drag-drop-area-overlay',
        template: "<div class=\"po-overlay-fixed\">\n  <po-upload-drag-drop-area\n    #DragDropAreaFixed\n    class=\"po-upload-drag-drop-area-overlay\"\n    [p-directory-compatible]=\"directoryCompatible\"\n    [p-disabled]=\"disabled\"\n    [p-literals]=\"literals\"\n    [p-overlay]=\"true\"\n  >\n  </po-upload-drag-drop-area>\n</div>\n",
        changeDetection: ChangeDetectionStrategy.OnPush
    }),
    __metadata("design:paramtypes", [Renderer2])
], PoUploadDragDropAreaOverlayComponent);
export { PoUploadDragDropAreaOverlayComponent };
