// @po-ui/ng-components/esm5/lib/components/po-breadcrumb/po-breadcrumb-favorite/po-breadcrumb-favorite.service.js
import { __decorate, __metadata } from "tslib";
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
/**
 * @docsPrivate
 *
 * @description
 *
 * Serviço utilizado para favoritar/desfavoritar uma url no componente po-breadcrumb.
 * A API deve retornar um objeto no formato { isFavorite: booleano, url: string }. Em ambos os métodos, GET e POST este
 * objeto deve ser retornado.
 * Ao alterar o status de favorito, a API deve retornar o objeto acima, com o novo status atualizado.
 * No método POST, existe a possibilidade de ser enviado parâmetros junto com a requisição. Neste caso, o objeto enviado
 * para a API terá o formato: { isFavorite: booleano, url: string, params: {} }, onde params é o objeto que pode ser enviado
 * junto com a requisição.
 *
 */
var PoBreadcrumbFavoriteService = /** @class */ (function () {
    function PoBreadcrumbFavoriteService(http) {
        this.http = http;
    }
    Object.defineProperty(PoBreadcrumbFavoriteService.prototype, "url", {
        get: function () {
            return this._url;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoBreadcrumbFavoriteService.prototype, "bodyParams", {
        get: function () {
            return this._bodyParams;
        },
        enumerable: true,
        configurable: true
    });
    PoBreadcrumbFavoriteService.prototype.configService = function (url, params, item) {
        this._url = url;
        this._bodyParams = {
            isFavorite: false,
            url: item.link,
            params: params
        };
    };
    PoBreadcrumbFavoriteService.prototype.sendStatusFavorite = function (status) {
        this.bodyParams['isFavorite'] = status;
        return this.http.post(this.url, this.bodyParams);
    };
    PoBreadcrumbFavoriteService.prototype.getFavorite = function () {
        return this.http.get(this.url, { responseType: 'json', params: { url: this.bodyParams['url'] } });
    };
    PoBreadcrumbFavoriteService.ctorParameters = function () { return [
        { type: HttpClient }
    ]; };
    PoBreadcrumbFavoriteService = __decorate([
        Injectable(),
        __metadata("design:paramtypes", [HttpClient])
    ], PoBreadcrumbFavoriteService);
    return PoBreadcrumbFavoriteService;
}());
export { PoBreadcrumbFavoriteService };
