// @po-ui/ng-components/esm2015/lib/components/po-tabs/po-tab/po-tab-base.component.js
import { __decorate, __metadata } from "tslib";
import { EventEmitter, Input, Output, Directive } from '@angular/core';
import { convertToBoolean, uuid } from '../../../utils/util';
/**
 * @description
 *
 * O componente `po-tab` renderiza uma aba na qual envolve um conteúdo HTML.
 *
 * Com este componente é possível atribuir um rótulo para auxiliar na identificação do conteúdo, ativar para que o mesmo seja exibido,
 * desabilitar para impossibilitar o acesso, bem como ocultar para indisponibilizar a aba.
 *
 * > Para controlar a navegação entre diversas abas, utilizar o componente [`po-tabs`](/documentation/po-tabs).
 */
let PoTabBaseComponent = class PoTabBaseComponent {
    constructor() {
        this._active = false;
        this._disabled = false;
        this._hide = false;
        // ID da aba
        this.id = uuid();
        /** Método disparado ao clicar na aba. */
        this.click = new EventEmitter();
    }
    /**
     * @optional
     *
     * @description
     *
     * Ativa a aba exibindo seu conteúdo.
     *
     * > Sugere-se utilizar na aba de conteúdo inicial.
     *
     * @default `false`
     */
    set active(active) {
        this._active = convertToBoolean(active);
        this.setDisplayOnActive();
    }
    get active() {
        return this._active;
    }
    /**
     * @optional
     *
     * @description
     *
     * Desabilita a aba.
     *
     * @default `false`
     */
    set disabled(disabled) {
        this._disabled = convertToBoolean(disabled);
    }
    get disabled() {
        return this._disabled;
    }
    /**
     * @optional
     *
     * @description
     *
     * Oculta a aba.
     *
     * @default `false`
     */
    set hide(hide) {
        this._hide = convertToBoolean(hide);
    }
    get hide() {
        return this._hide;
    }
};
__decorate([
    Input('p-active'),
    __metadata("design:type", Boolean),
    __metadata("design:paramtypes", [Boolean])
], PoTabBaseComponent.prototype, "active", null);
__decorate([
    Input('p-disabled'),
    __metadata("design:type", Boolean),
    __metadata("design:paramtypes", [Boolean])
], PoTabBaseComponent.prototype, "disabled", null);
__decorate([
    Input('p-hide'),
    __metadata("design:type", Boolean),
    __metadata("design:paramtypes", [Boolean])
], PoTabBaseComponent.prototype, "hide", null);
__decorate([
    Input('p-label'),
    __metadata("design:type", String)
], PoTabBaseComponent.prototype, "label", void 0);
__decorate([
    Output('p-click'),
    __metadata("design:type", Object)
], PoTabBaseComponent.prototype, "click", void 0);
PoTabBaseComponent = __decorate([
    Directive()
], PoTabBaseComponent);
export { PoTabBaseComponent };
