// @po-ui/ng-components/esm5/lib/components/po-widget/po-widget.component.js
import { __decorate, __extends, __metadata } from "tslib";
import { Component, OnInit, ViewContainerRef } from '@angular/core';
import { PoWidgetBaseComponent } from './po-widget-base.component';
/**
 * @docsExtends PoWidgetBaseComponent
 *
 * @example
 *
 * <example name="po-widget-basic" title="PO Widget Basic">
 *  <file name="sample-po-widget-basic/sample-po-widget-basic.component.html"> </file>
 *  <file name="sample-po-widget-basic/sample-po-widget-basic.component.ts"> </file>
 * </example>
 *
 * <example name="po-widget-labs" title="PO Widget Labs">
 *  <file name="sample-po-widget-labs/sample-po-widget-labs.component.html"> </file>
 *  <file name="sample-po-widget-labs/sample-po-widget-labs.component.ts"> </file>
 * </example>
 *
 * <example name="po-widget-finance-dashboard" title="PO Widget - Finance dashboard">
 *  <file name="sample-po-widget-finance-dashboard/sample-po-widget-finance-dashboard.component.html"> </file>
 *  <file name="sample-po-widget-finance-dashboard/sample-po-widget-finance-dashboard.component.ts"> </file>
 * </example>
 *
 * <example name="po-widget-card" title="PO Widget - Card">
 *  <file name="sample-po-widget-card/sample-po-widget-card.component.html"> </file>
 *  <file name="sample-po-widget-card/sample-po-widget-card.component.ts"> </file>
 * </example>
 *
 */
var PoWidgetComponent = /** @class */ (function (_super) {
    __extends(PoWidgetComponent, _super);
    function PoWidgetComponent(viewRef) {
        return _super.call(this) || this;
    }
    Object.defineProperty(PoWidgetComponent.prototype, "showTitleAction", {
        get: function () {
            return !!this.titleAction.observers[0];
        },
        enumerable: true,
        configurable: true
    });
    PoWidgetComponent.prototype.ngOnInit = function () {
        this.setHeight(this.height);
    };
    PoWidgetComponent.prototype.hasTitleHelpOrSetting = function () {
        return !!this.title || !!this.help || !!this.setting.observers[0];
    };
    PoWidgetComponent.prototype.onClick = function (event) {
        if (!this.disabled) {
            this.click.emit(event);
        }
    };
    PoWidgetComponent.prototype.openHelp = function (event) {
        if (!this.disabled) {
            event.stopPropagation();
            window.open(this.help, '_blank');
        }
    };
    PoWidgetComponent.prototype.runPrimaryAction = function (event) {
        if (!this.disabled) {
            event.stopPropagation();
            this.primaryAction.emit();
        }
    };
    PoWidgetComponent.prototype.runSecondaryAction = function (event) {
        if (!this.disabled) {
            event.stopPropagation();
            this.secondaryAction.emit();
        }
    };
    PoWidgetComponent.prototype.runTitleAction = function (event) {
        if (!this.disabled) {
            event.stopPropagation();
            this.titleAction.emit();
        }
    };
    PoWidgetComponent.prototype.setHeight = function (height) {
        if (height) {
            var bodyHeight = height;
            var hasSettingOrHelp = this.setting.observers.length > 0 || this.help;
            var footerBorder = 1;
            var footerHeight = 40;
            var settingHeight = 37;
            var shadowHeight = 2;
            var titleHeight = 50;
            if (this.noShadow) {
                bodyHeight -= shadowHeight;
            }
            if (hasSettingOrHelp && !this.title) {
                bodyHeight -= settingHeight;
            }
            if (this.title) {
                bodyHeight -= titleHeight;
            }
            if (this.primaryLabel) {
                bodyHeight -= footerHeight + footerBorder;
            }
            this.containerHeight = bodyHeight + "px";
        }
        else {
            this.containerHeight = "auto";
        }
    };
    PoWidgetComponent.prototype.settingOutput = function (event) {
        if (!this.disabled) {
            event.stopPropagation();
            this.setting.emit();
        }
    };
    PoWidgetComponent.ctorParameters = function () { return [
        { type: ViewContainerRef }
    ]; };
    PoWidgetComponent = __decorate([
        Component({
            selector: 'po-widget',
            template: "<div\n  [class.po-clickable]=\"click.observers.length\"\n  [class.po-widget]=\"!primary\"\n  [class.po-widget-disabled]=\"disabled\"\n  [class.po-widget-primary]=\"primary\"\n  [class.po-widget-no-shadow]=\"noShadow\"\n  [style.background-image]=\"background ? 'url(' + background + ')' : undefined\"\n  (click)=\"onClick($event)\"\n>\n  <div *ngIf=\"hasTitleHelpOrSetting()\" class=\"po-widget-header\">\n    <span *ngIf=\"showTitleAction; else noTitleAction\" class=\"po-widget-title-action\" (click)=\"runTitleAction($event)\">\n      {{ title }}\n    </span>\n    <ng-template #noTitleAction>{{ title }}</ng-template>\n\n    <div *ngIf=\"setting.observers[0] || help\" class=\"po-pull-right\">\n      <span\n        id=\"settingAction\"\n        *ngIf=\"setting.observers[0]\"\n        class=\"po-clickable po-icon po-icon-settings\"\n        (click)=\"settingOutput($event)\"\n      >\n      </span>\n      <span id=\"helpLink\" *ngIf=\"help\" class=\"po-clickable po-icon po-icon-help\" (click)=\"openHelp($event)\"> </span>\n    </div>\n  </div>\n\n  <po-container p-no-border [p-height]=\"containerHeight\">\n    <div class=\"po-widget-body\">\n      <ng-content></ng-content>\n    </div>\n  </po-container>\n\n  <div *ngIf=\"primaryLabel\" class=\"po-widget-footer\">\n    <div class=\"po-widget-xl\" *ngIf=\"!secondaryLabel\">\n      <a class=\"po-widget-action\" id=\"primaryAct\" (click)=\"runPrimaryAction($event)\">{{ primaryLabel }}</a>\n    </div>\n\n    <div class=\"po-widget-md\" *ngIf=\"secondaryLabel\">\n      <a class=\"po-widget-action\" id=\"primaryAct\" (click)=\"runPrimaryAction($event)\">{{ primaryLabel }}</a>\n    </div>\n    <div class=\"po-widget-md\" *ngIf=\"secondaryLabel\">\n      <a class=\"po-widget-action\" id=\"secondaryAct\" (click)=\"runSecondaryAction($event)\">{{ secondaryLabel }}</a>\n    </div>\n  </div>\n</div>\n"
        }),
        __metadata("design:paramtypes", [ViewContainerRef])
    ], PoWidgetComponent);
    return PoWidgetComponent;
}(PoWidgetBaseComponent));
export { PoWidgetComponent };
