// @po-ui/ng-components/esm5/lib/components/po-accordion/po-accordion.component.js
import { __decorate, __extends, __metadata } from "tslib";
import { Component, ContentChildren, QueryList } from '@angular/core';
import { PoAccordionBaseComponent } from './po-accordion-base.component';
import { PoAccordionItemComponent } from './po-accordion-item/po-accordion-item.component';
import { PoAccordionService } from './services/po-accordion.service';
/**
 * @docsExtends PoAccordionBaseComponent
 *
 * @example
 *
 * <example name="po-accordion-basic" title="PO Accordion Basic" >
 *  <file name="sample-po-accordion-basic/sample-po-accordion-basic.component.html"> </file>
 *  <file name="sample-po-accordion-basic/sample-po-accordion-basic.component.ts"> </file>
 *  <file name="sample-po-accordion-basic/sample-po-accordion-basic.component.e2e-spec.ts"> </file>
 *  <file name="sample-po-accordion-basic/sample-po-accordion-basic.component.po.ts"> </file>
 * </example>
 *
 * <example name="po-accordion-labs" title="PO Accordion Labs" >
 *  <file name="sample-po-accordion-labs/sample-po-accordion-labs.component.html"> </file>
 *  <file name="sample-po-accordion-labs/sample-po-accordion-labs.component.ts"> </file>
 * </example>
 *
 * <example name="po-accordion-faq" title="PO Accordion - FAQs" >
 *  <file name="sample-po-accordion-faq/sample-po-accordion-faq.component.html"> </file>
 *  <file name="sample-po-accordion-faq/sample-po-accordion-faq.component.ts"> </file>
 * </example>
 */
var PoAccordionComponent = /** @class */ (function (_super) {
    __extends(PoAccordionComponent, _super);
    function PoAccordionComponent(accordionService) {
        var _this = _super.call(this) || this;
        _this.accordionService = accordionService;
        _this.receiveFromChildAccordionSubscription();
        return _this;
    }
    PoAccordionComponent.prototype.ngOnDestroy = function () {
        this.accordionServiceSubscription.unsubscribe();
    };
    PoAccordionComponent.prototype.headerToggle = function (event, poAccordionItem) {
        poAccordionItem.expanded = event;
        this.toggle(poAccordionItem);
    };
    PoAccordionComponent.prototype.receiveFromChildAccordionSubscription = function () {
        var _this = this;
        this.accordionServiceSubscription = this.accordionService
            .receiveFromChildAccordionClicked()
            .subscribe(function (poAccordionItem) { return _this.toggle(poAccordionItem); });
    };
    PoAccordionComponent.prototype.toggle = function (poAccordionItem) {
        var isCurrentAccordionCollapsed = !poAccordionItem.expanded;
        if (isCurrentAccordionCollapsed) {
            this.expandedActiveAccordionItem = null;
            return;
        }
        if (this.expandedActiveAccordionItem) {
            this.expandedActiveAccordionItem.collapse();
        }
        this.expandedActiveAccordionItem = poAccordionItem;
    };
    PoAccordionComponent.ctorParameters = function () { return [
        { type: PoAccordionService }
    ]; };
    __decorate([
        ContentChildren(PoAccordionItemComponent),
        __metadata("design:type", QueryList)
    ], PoAccordionComponent.prototype, "poAccordionItems", void 0);
    PoAccordionComponent = __decorate([
        Component({
            selector: 'po-accordion',
            template: "<div class=\"po-accordion po-container po-container-no-padding\">\n  <section\n    *ngFor=\"let poAccordionItem of poAccordionItems\"\n    class=\"po-accordion-item\"\n    [class.po-accordion-item-active]=\"poAccordionItem.expanded\"\n  >\n    <po-accordion-item-header\n      [p-expanded]=\"poAccordionItem.expanded\"\n      [p-label]=\"poAccordionItem.label\"\n      (p-toggle)=\"headerToggle($event, poAccordionItem)\"\n    >\n    </po-accordion-item-header>\n\n    <po-accordion-item-body [p-expanded]=\"poAccordionItem.expanded\">\n      <ng-container *ngTemplateOutlet=\"poAccordionItem.templateRef\"></ng-container>\n    </po-accordion-item-body>\n  </section>\n</div>\n",
            providers: [PoAccordionService]
        }),
        __metadata("design:paramtypes", [PoAccordionService])
    ], PoAccordionComponent);
    return PoAccordionComponent;
}(PoAccordionBaseComponent));
export { PoAccordionComponent };
