// @po-ui/ng-components/esm5/lib/components/po-field/po-upload/po-upload-file-restrictions/po-upload-file-restrictions.component.js
import { __assign, __decorate, __metadata } from "tslib";
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, Input, OnInit } from '@angular/core';
import { browserLanguage, formatBytes, poLocaleDefault } from '../../../../utils/util';
import { poUploadLiteralsDefault } from '../po-upload-base.component';
var PoUploadFileRestrictionsComponent = /** @class */ (function () {
    function PoUploadFileRestrictionsComponent(changeDetector) {
        this.changeDetector = changeDetector;
    }
    Object.defineProperty(PoUploadFileRestrictionsComponent.prototype, "allowedExtensions", {
        get: function () {
            return this._allowedExtensions;
        },
        set: function (value) {
            this._allowedExtensions = this.formatAllowedExtensions(value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoUploadFileRestrictionsComponent.prototype, "maxFileSize", {
        get: function () {
            return this._maxFileSize;
        },
        set: function (value) {
            this._maxFileSize = formatBytes(value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoUploadFileRestrictionsComponent.prototype, "minFileSize", {
        get: function () {
            return this._minFileSize;
        },
        set: function (value) {
            this._minFileSize = formatBytes(value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoUploadFileRestrictionsComponent.prototype, "language", {
        get: function () {
            return browserLanguage();
        },
        enumerable: true,
        configurable: true
    });
    PoUploadFileRestrictionsComponent.prototype.ngOnInit = function () {
        this.setLiterals();
    };
    PoUploadFileRestrictionsComponent.prototype.formatAllowedExtensions = function (allowedExtensions) {
        var conjunction = { 'pt': 'e', 'en': 'and', 'es': 'y' };
        return allowedExtensions
            ? allowedExtensions
                .join(', ')
                .toUpperCase()
                .replace(/,(?=[^,]*$)/, " " + conjunction[this.language])
            : undefined;
    };
    PoUploadFileRestrictionsComponent.prototype.setLiterals = function () {
        this.literals = __assign(__assign({}, poUploadLiteralsDefault[poLocaleDefault]), poUploadLiteralsDefault[this.language]);
        this.changeDetector.detectChanges();
    };
    PoUploadFileRestrictionsComponent.ctorParameters = function () { return [
        { type: ChangeDetectorRef }
    ]; };
    __decorate([
        Input('p-allowed-extensions'),
        __metadata("design:type", String),
        __metadata("design:paramtypes", [Object])
    ], PoUploadFileRestrictionsComponent.prototype, "allowedExtensions", null);
    __decorate([
        Input('p-max-files'),
        __metadata("design:type", Number)
    ], PoUploadFileRestrictionsComponent.prototype, "maxFiles", void 0);
    __decorate([
        Input('p-max-file-size'),
        __metadata("design:type", String),
        __metadata("design:paramtypes", [Object])
    ], PoUploadFileRestrictionsComponent.prototype, "maxFileSize", null);
    __decorate([
        Input('p-min-file-size'),
        __metadata("design:type", String),
        __metadata("design:paramtypes", [Object])
    ], PoUploadFileRestrictionsComponent.prototype, "minFileSize", null);
    PoUploadFileRestrictionsComponent = __decorate([
        Component({
            selector: 'po-upload-file-restrictions',
            template: "<p class=\"po-font-text-small\" *ngIf=\"maxFiles > 1\">\n  {{ literals?.numberOfFilesAllowed | poI18n: maxFiles }}\n</p>\n\n<p class=\"po-font-text-small\" *ngIf=\"allowedExtensions\">\n  {{ literals?.allowedFormats | poI18n: allowedExtensions }}\n</p>\n\n<p class=\"po-font-text-small\" *ngIf=\"minFileSize || maxFileSize\">\n  <span>{{ literals?.allowedSizes }}</span>\n\n  <span *ngIf=\"minFileSize && maxFileSize\">\n    {{ literals?.allowedFileSizeRange | poI18n: [minFileSize, maxFileSize] }}\n  </span>\n\n  <span *ngIf=\"minFileSize && !maxFileSize\">\n    {{ literals?.minFileSizeAllowed | poI18n: minFileSize }}\n  </span>\n\n  <span *ngIf=\"maxFileSize && !minFileSize\">\n    {{ literals?.maxFileSizeAllowed | poI18n: maxFileSize }}\n  </span>\n</p>\n",
            changeDetection: ChangeDetectionStrategy.OnPush
        }),
        __metadata("design:paramtypes", [ChangeDetectorRef])
    ], PoUploadFileRestrictionsComponent);
    return PoUploadFileRestrictionsComponent;
}());
export { PoUploadFileRestrictionsComponent };
