// @po-ui/ng-components/esm5/lib/components/po-menu/services/po-menu-items.service.js
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
/**
 * @docsPrivate
 *
 * @description
 *
 * Serviço que implementa a comunicação entre os items do po-menu.
 */
var PoMenuItemsService = /** @class */ (function () {
    function PoMenuItemsService() {
        this.subjectParent = new Subject();
        this.subjectChild = new Subject();
    }
    // Envia informações do click do po-menu-item para o po-menu
    PoMenuItemsService.prototype.sendToParentMenuClicked = function (menu) {
        this.subjectChild.next(menu);
    };
    // Recebe do po-menu-item sua informação de click.
    PoMenuItemsService.prototype.receiveFromChildMenuClicked = function () {
        return this.subjectChild.asObservable();
    };
    // Envia para os po-menu-item a resposta do processamento de click de um po-menu-item.
    PoMenuItemsService.prototype.sendToChildMenuClicked = function (menu) {
        this.subjectParent.next(menu);
    };
    // Recebe do po-menu as informações processadas do click de um po-menu-item.
    PoMenuItemsService.prototype.receiveFromParentMenuClicked = function () {
        return this.subjectParent.asObservable();
    };
    PoMenuItemsService = __decorate([
        Injectable()
    ], PoMenuItemsService);
    return PoMenuItemsService;
}());
export { PoMenuItemsService };
