// @po-ui/ng-components/esm2015/lib/components/po-field/po-field-container/po-field-container.component.js
import { __decorate, __metadata } from "tslib";
import { Component, Input } from '@angular/core';
import { convertToBoolean, getShortBrowserLanguage } from '../../../utils/util';
/**
 * @docsPrivate
 *
 * Componente de uso interno, responsável por atribuir uma label para o campo
 */
let PoFieldContainerComponent = class PoFieldContainerComponent {
    constructor() {
        this._optional = false;
    }
    /** Indica se o campo será opcional. */
    set optional(value) {
        this._optional = convertToBoolean(value);
    }
    get optional() {
        return this._optional;
    }
    getOptionalText() {
        const browserLanguage = getShortBrowserLanguage();
        const optional = {
            pt: '(Opcional)',
            en: '(Optional)',
            es: '(Opcional)'
        };
        return optional[browserLanguage];
    }
};
__decorate([
    Input('p-label'),
    __metadata("design:type", String)
], PoFieldContainerComponent.prototype, "label", void 0);
__decorate([
    Input('p-help'),
    __metadata("design:type", String)
], PoFieldContainerComponent.prototype, "help", void 0);
__decorate([
    Input('p-optional'),
    __metadata("design:type", Boolean),
    __metadata("design:paramtypes", [Boolean])
], PoFieldContainerComponent.prototype, "optional", null);
PoFieldContainerComponent = __decorate([
    Component({
        selector: 'po-field-container',
        template: "<div class=\"po-field-container\">\n  <div class=\"po-field-container-title\">\n    <span class=\"po-field-title\">{{ label }}</span>\n    <span class=\"po-field-optional\" *ngIf=\"optional && (label || help)\">{{ getOptionalText() }}</span>\n  </div>\n\n  <div class=\"po-field-help\" *ngIf=\"help\">{{ help }}</div>\n\n  <ng-content></ng-content>\n</div>\n"
    })
], PoFieldContainerComponent);
export { PoFieldContainerComponent };
