// @po-ui/ng-components/esm2015/lib/components/po-navbar/po-navbar-items/po-navbar-item/po-navbar-item.component.js
import { __decorate, __metadata } from "tslib";
import { Component, Input, EventEmitter, Output } from '@angular/core';
import { isExternalLink } from '../../../../utils/util';
let PoNavbarItemComponent = class PoNavbarItemComponent {
    constructor() {
        this.click = new EventEmitter();
    }
    get type() {
        return isExternalLink(this.link) ? 'externalLink' : 'internalLink';
    }
    itemClick(label, link) {
        if (this.action) {
            this.action({ label, link });
        }
        this.click.emit();
    }
};
__decorate([
    Input('p-action'),
    __metadata("design:type", Function)
], PoNavbarItemComponent.prototype, "action", void 0);
__decorate([
    Input('p-clickable'),
    __metadata("design:type", Boolean)
], PoNavbarItemComponent.prototype, "clickable", void 0);
__decorate([
    Input('p-label'),
    __metadata("design:type", String)
], PoNavbarItemComponent.prototype, "label", void 0);
__decorate([
    Input('p-link'),
    __metadata("design:type", String)
], PoNavbarItemComponent.prototype, "link", void 0);
__decorate([
    Output('p-click'),
    __metadata("design:type", EventEmitter)
], PoNavbarItemComponent.prototype, "click", void 0);
PoNavbarItemComponent = __decorate([
    Component({
        selector: 'po-navbar-item',
        template: "<ng-container [ngSwitch]=\"type\">\n  <a\n    *ngSwitchCase=\"'externalLink'\"\n    class=\"po-navbar-item-link\"\n    [class.po-clickable]=\"clickable\"\n    [href]=\"link\"\n    target=\"_blank\"\n    (click)=\"itemClick()\"\n  >\n    {{ label }}\n  </a>\n\n  <a\n    *ngSwitchCase=\"'internalLink'\"\n    class=\"po-navbar-item-link\"\n    [class.po-clickable]=\"clickable\"\n    [routerLink]=\"link\"\n    (click)=\"itemClick(label, link)\"\n  >\n    {{ label }}\n  </a>\n</ng-container>\n"
    })
], PoNavbarItemComponent);
export { PoNavbarItemComponent };
