// @po-ui/ng-components/esm5/lib/components/po-menu-panel/po-menu-panel-base.component.js
import { __decorate, __metadata } from "tslib";
import { Input, Directive } from '@angular/core';
import { v4 as uuid } from 'uuid';
import { isExternalLink } from '../../utils/util';
/**
 * @description
 *
 * Este é um componente de menu lateral composto apenas por ícones e com um nível, utilizado para navegação
 * em páginas internas, externas da aplicação ou aciona uma ação.
 *
 * O componente `po-menu-panel` recebe uma lista de objetos do tipo `MenuPanelItem` com as informações dos
 * itens de menu como textos, links para redirecionamento, ações e ícones.
 */
var PoMenuPanelBaseComponent = /** @class */ (function () {
    function PoMenuPanelBaseComponent() {
    }
    Object.defineProperty(PoMenuPanelBaseComponent.prototype, "menus", {
        get: function () {
            return this._menus;
        },
        /** Lista dos itens do `po-menu-panel`. Se o valor estiver indefinido ou inválido, será inicializado como um array vazio. */
        set: function (menus) {
            this._menus = Array.isArray(menus) ? menus : [];
            this.setMenuExtraProperties(this._menus);
            this.validateMenus(this._menus);
        },
        enumerable: true,
        configurable: true
    });
    PoMenuPanelBaseComponent.prototype.setMenuExtraProperties = function (menus) {
        var _this = this;
        menus.forEach(function (menuItem) { return _this.setMenuItemProperties(menuItem); });
    };
    PoMenuPanelBaseComponent.prototype.setMenuItemProperties = function (menuItem) {
        menuItem.id = menuItem.id || uuid();
        menuItem.type = this.setMenuType(menuItem);
    };
    PoMenuPanelBaseComponent.prototype.setMenuType = function (menuItem) {
        if (!menuItem.link) {
            return 'noLink';
        }
        if (isExternalLink(menuItem.link)) {
            return 'externalLink';
        }
        return 'internalLink';
    };
    PoMenuPanelBaseComponent.prototype.validateMenu = function (menuItem) {
        if (!menuItem.label) {
            throw new Error('O atributo PoMenuPanelItem.label não pode ser vazio.');
        }
        if (!menuItem.icon) {
            throw new Error('O atributo PoMenuPanelItem.icon não pode ser vazio.');
        }
    };
    PoMenuPanelBaseComponent.prototype.validateMenus = function (menus) {
        var _this = this;
        menus.forEach(function (menu) { return _this.validateMenu(menu); });
    };
    __decorate([
        Input('p-menus'),
        __metadata("design:type", Array),
        __metadata("design:paramtypes", [Array])
    ], PoMenuPanelBaseComponent.prototype, "menus", null);
    PoMenuPanelBaseComponent = __decorate([
        Directive()
    ], PoMenuPanelBaseComponent);
    return PoMenuPanelBaseComponent;
}());
export { PoMenuPanelBaseComponent };
