// @po-ui/ng-components/esm5/lib/components/po-toolbar/po-toolbar.module.js
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { PoAvatarModule } from './../po-avatar/po-avatar.module';
import { PoPopupModule } from '../po-popup/po-popup.module';
import { PoToolbarActionsComponent } from './po-toolbar-actions/po-toolbar-actions.component';
import { PoToolbarComponent } from './po-toolbar.component';
import { PoToolbarNotificationComponent } from './po-toolbar-notification/po-toolbar-notification.component';
import { PoToolbarProfileComponent } from './po-toolbar-profile/po-toolbar-profile.component';
/**
 * @description
 *
 * Módulo do componente po-toolbar
 *
 */
var PoToolbarModule = /** @class */ (function () {
    function PoToolbarModule() {
    }
    PoToolbarModule = __decorate([
        NgModule({
            imports: [CommonModule, PoAvatarModule, PoPopupModule],
            declarations: [
                PoToolbarActionsComponent,
                PoToolbarComponent,
                PoToolbarNotificationComponent,
                PoToolbarProfileComponent
            ],
            exports: [PoToolbarComponent],
            providers: [],
            schemas: []
        })
    ], PoToolbarModule);
    return PoToolbarModule;
}());
export { PoToolbarModule };
