// @po-ui/ng-components/esm5/lib/components/po-page/po-page-detail/po-page-detail.component.js
import { __decorate, __extends } from "tslib";
import { Component } from '@angular/core';
import { PoPageDetailBaseComponent } from './po-page-detail-base.component';
/**
 * @docsExtends PoPageDetailBaseComponent
 *
 * @example
 *
 * <example name="po-page-detail-basic" title="PO Page Detail Basic">
 *  <file name="sample-po-page-detail-basic/sample-po-page-detail-basic.component.html"> </file>
 *  <file name="sample-po-page-detail-basic/sample-po-page-detail-basic.component.ts"> </file>
 * </example>
 *
 * <example name="po-page-detail-labs" title="PO Page Detail Labs">
 *  <file name="sample-po-page-detail-labs/sample-po-page-detail-labs.component.html"> </file>
 *  <file name="sample-po-page-detail-labs/sample-po-page-detail-labs.component.ts"> </file>
 * </example>
 *
 * <example name="po-page-detail-user" title="PO Page Detail - User">
 *  <file name="sample-po-page-detail-user/sample-po-page-detail-user.component.html"> </file>
 *  <file name="sample-po-page-detail-user/sample-po-page-detail-user.component.ts"> </file>
 * </example>
 */
var PoPageDetailComponent = /** @class */ (function (_super) {
    __extends(PoPageDetailComponent, _super);
    function PoPageDetailComponent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    PoPageDetailComponent.prototype.hasAnyAction = function () {
        return this.hasEvent('back') || this.hasEvent('edit') || this.hasEvent('remove');
    };
    PoPageDetailComponent.prototype.hasEditFn = function (property) {
        if (property === 'icon') {
            return this.hasEvent('edit') ? '' : 'po-icon-delete';
        }
        else if (property === 'type') {
            return this.hasEvent('edit') ? 'default' : 'primary';
        }
        else {
            return '';
        }
    };
    PoPageDetailComponent.prototype.hasEditOrRemoveFn = function (property) {
        if (property === 'icon') {
            return this.hasEvent('edit') || this.hasEvent('remove') ? '' : 'po-icon-arrow-left';
        }
        else if (property === 'type') {
            return this.hasEvent('edit') || this.hasEvent('remove') ? 'default' : 'primary';
        }
        else {
            return '';
        }
    };
    PoPageDetailComponent.prototype.hasEvent = function (event) {
        return !!this[event].observers.length;
    };
    PoPageDetailComponent.prototype.hasPageHeader = function () {
        return !!(this.title || this.hasAnyAction() || (this.breadcrumb && this.breadcrumb.items.length));
    };
    PoPageDetailComponent = __decorate([
        Component({
            selector: 'po-page-detail',
            template: "<po-page>\n  <!-- HEADER -->\n  <po-page-header *ngIf=\"hasPageHeader()\" [p-breadcrumb]=\"breadcrumb\" [p-title]=\"title\">\n    <!-- OPERATIONS -->\n    <div *ngIf=\"hasAnyAction()\" class=\"po-page-header-actions\">\n      <po-button\n        *ngIf=\"hasEvent('back')\"\n        [p-icon]=\"hasEditOrRemoveFn('icon')\"\n        [p-label]=\"literals.back\"\n        [p-type]=\"hasEditOrRemoveFn('type')\"\n        (p-click)=\"back.emit()\"\n      >\n      </po-button>\n\n      <po-button\n        *ngIf=\"hasEvent('remove')\"\n        [p-icon]=\"hasEditFn('icon')\"\n        [p-label]=\"literals.remove\"\n        [p-type]=\"hasEditFn('type')\"\n        (p-click)=\"remove.emit()\"\n      >\n      </po-button>\n\n      <po-button\n        *ngIf=\"hasEvent('edit')\"\n        p-icon=\"po-icon-edit\"\n        p-type=\"primary\"\n        [p-label]=\"literals.edit\"\n        (p-click)=\"edit.emit()\"\n      >\n      </po-button>\n    </div>\n  </po-page-header>\n\n  <!-- CONTENT -->\n  <po-page-content>\n    <ng-content> </ng-content>\n  </po-page-content>\n</po-page>\n"
        })
    ], PoPageDetailComponent);
    return PoPageDetailComponent;
}(PoPageDetailBaseComponent));
export { PoPageDetailComponent };
