// @po-ui/ng-components/esm5/lib/components/po-field/po-upload/po-upload-base.service.js
import { __decorate, __metadata } from "tslib";
import { Injectable } from '@angular/core';
import { HttpClient, HttpEventType, HttpErrorResponse, HttpRequest, HttpResponse } from '@angular/common/http';
var PoUploadBaseService = /** @class */ (function () {
    function PoUploadBaseService(http) {
        this.http = http;
        this.formField = 'files';
        this.requests = [];
    }
    /**
     * Método responsável por enviar os arquivos ao servidor, conforme o parâmetro URL.
     *
     * @param url URL da requisição a ser efetuada.
     * @param files Arquivos a serem enviados.
     * @param tOnUpload Função a ser executada quando o arquivo for enviado ao servidor.
     * @param uploadCallback Função que será executada enquanto os arquivos estiverem sendo enviados.
     * @param successCallback Função a ser executada quando a requisição for efetuada com sucesso.
     * @param errorCallback Função a ser executada quando a requisição foi efetuada com sucesso.
     */
    PoUploadBaseService.prototype.upload = function (url, files, tOnUpload, uploadCallback, successCallback, errorCallback) {
        var filesLength = files.length;
        var uploadEvent = {
            data: {},
            file: null
        };
        for (var i = 0; i < filesLength; i++) {
            var formData = new FormData();
            var file = files[i];
            var fileName = file.rawFile.name;
            formData.append(this.formField, file.rawFile, fileName);
            // Função upload, onde o desenvolvedor pode enviar dados para a requisição.
            if (tOnUpload) {
                uploadEvent['file'] = file;
                tOnUpload.emit(uploadEvent);
                formData.append('data', JSON.stringify(uploadEvent.data));
            }
            this.sendFile(url, file, formData, uploadCallback, successCallback, errorCallback);
        }
    };
    PoUploadBaseService.prototype.sendFile = function (url, file, formData, uploadCallback, successCallback, errorCallback) {
        var _this = this;
        var request = this.getRequest(url, formData).subscribe(function (event) {
            if (event.type === HttpEventType.UploadProgress) {
                _this.addRequest(file, request);
                var percentDone = Math.round((100 * event.loaded) / event.total);
                uploadCallback(file, percentDone);
            }
            else if (event instanceof HttpResponse) {
                // Sucesso, arquivos enviados.
                successCallback(file, event);
            }
        }, function (err) {
            errorCallback(file, err);
        });
    };
    PoUploadBaseService.prototype.getRequest = function (url, formData) {
        var req = new HttpRequest('POST', url, formData, {
            reportProgress: true
        });
        return this.http.request(req);
    };
    PoUploadBaseService.prototype.stopRequestByFile = function (file, callback) {
        var requestObj = this.requests.find(function (req) {
            return req.file.uid === file.uid;
        });
        if (requestObj) {
            var request = requestObj.request;
            request.unsubscribe();
            this.removeRequest(requestObj);
            callback();
        }
    };
    PoUploadBaseService.prototype.removeRequest = function (requestObj) {
        var index = this.requests.indexOf(requestObj);
        this.requests.splice(index, 1);
    };
    PoUploadBaseService.prototype.addRequest = function (file, request) {
        var hasRequest = this.requests.some(function (req) {
            return req.file.uid === file.uid;
        });
        if (!hasRequest) {
            this.requests.push({ file: file, request: request });
        }
    };
    PoUploadBaseService.ctorParameters = function () { return [
        { type: HttpClient }
    ]; };
    PoUploadBaseService = __decorate([
        Injectable(),
        __metadata("design:paramtypes", [HttpClient])
    ], PoUploadBaseService);
    return PoUploadBaseService;
}());
export { PoUploadBaseService };
