// @po-ui/ng-components/esm5/lib/components/po-field/po-upload/po-upload-file.js
import { PoUploadStatus } from './po-upload-status.enum';
// Classe responsável pelo arquivo do PO Upload
var PoUploadFile = /** @class */ (function () {
    function PoUploadFile(file) {
        if (file) {
            this.name = file.name;
            this.displayName = file.name + " - " + this.getFileSize(file.size);
            this.extension = this.getExtension(file.name);
            this.size = file.size;
            this.rawFile = file;
            this.uid = this.generateUUID();
            this.status = PoUploadStatus.None;
        }
    }
    PoUploadFile.prototype.getExtension = function (value) {
        if (value) {
            var extension = value.substr(value.lastIndexOf('.'));
            return extension.toLowerCase();
        }
        return '';
    };
    PoUploadFile.prototype.generateUUID = function () {
        function s4() {
            return Math.floor((1 + Math.random()) * 0x10000)
                .toString(16)
                .substring(1);
        }
        return s4() + s4() + '-' + s4() + '-' + s4() + '-' + s4() + '-' + s4() + s4() + s4();
    };
    // Retorna o tamanho do arquivo em KBytes.
    PoUploadFile.prototype.getFileSize = function (size) {
        var kbSize = 0;
        if (size) {
            kbSize = Math.ceil(size / 1024);
        }
        return kbSize + " KB";
    };
    return PoUploadFile;
}());
export { PoUploadFile };
