// @po-ui/ng-components/esm5/lib/interceptors/po-http-interceptor/po-http-interceptor-detail/po-http-interceptor-detail.component.js
import { __decorate, __metadata } from "tslib";
import { Component, EventEmitter, ViewChild } from '@angular/core';
import { getShortBrowserLanguage } from '../../../utils/util';
import { PoModalComponent } from '../../../components/po-modal/po-modal.component';
import { poHttpInterceptorDetailLiteralsDefault } from './po-http-interceptor-detail-literals.interface';
export var colors = { success: 'color-11', error: 'color-07', warning: 'color-08', info: '' };
/**
 * @docsPrivate
 *
 * @description
 *
 * Componente para a modal de detalhes exibida pelo interceptor
 */
var PoHttpInterceptorDetailComponent = /** @class */ (function () {
    function PoHttpInterceptorDetailComponent() {
        var _this = this;
        this.closed = new EventEmitter();
        this.details = [];
        this.language = getShortBrowserLanguage();
        this.literals = poHttpInterceptorDetailLiteralsDefault[this.language];
        this.primaryAction = {
            action: function () { return _this.close(); },
            label: this.literals.closeButton
        };
    }
    Object.defineProperty(PoHttpInterceptorDetailComponent.prototype, "detail", {
        set: function (details) {
            if (details && details.length) {
                this.details = this.filterByValidDetails(details);
            }
            this.title = this.formatTitle(this.details);
        },
        enumerable: true,
        configurable: true
    });
    PoHttpInterceptorDetailComponent.prototype.close = function () {
        this.modal.close();
        this.closed.emit();
    };
    PoHttpInterceptorDetailComponent.prototype.formatDetailItemTitle = function (detail) {
        return detail.code ? detail.code + " - " + detail.message : detail.message;
    };
    PoHttpInterceptorDetailComponent.prototype.open = function () {
        this.modal.open();
    };
    PoHttpInterceptorDetailComponent.prototype.typeColor = function (type) {
        return colors[type];
    };
    PoHttpInterceptorDetailComponent.prototype.typeValue = function (type) {
        return poHttpInterceptorDetailLiteralsDefault[this.language][type] || type;
    };
    PoHttpInterceptorDetailComponent.prototype.addValidDetail = function (newDetails, detail) {
        return detail.message ? newDetails.concat(this.getValidDetailProperties(detail)) : newDetails;
    };
    PoHttpInterceptorDetailComponent.prototype.getValidDetailProperties = function (_a) {
        var code = _a.code, message = _a.message, detailedMessage = _a.detailedMessage, type = _a.type;
        return { code: code, message: message, detailedMessage: detailedMessage, type: type };
    };
    PoHttpInterceptorDetailComponent.prototype.filterByValidDetails = function (details) {
        var _this = this;
        return details.reduce(function (newDetails, detail) { return _this.addValidDetail(newDetails, detail); }, []);
    };
    PoHttpInterceptorDetailComponent.prototype.formatTitle = function (details) {
        return details.length > 1 ? this.literals.details + " (" + details.length + ")" : this.literals.detail;
    };
    __decorate([
        ViewChild(PoModalComponent, { static: true }),
        __metadata("design:type", PoModalComponent)
    ], PoHttpInterceptorDetailComponent.prototype, "modal", void 0);
    PoHttpInterceptorDetailComponent = __decorate([
        Component({
            selector: 'po-http-interceptor-detail',
            template: "<po-modal p-hide-close p-size=\"lg\" [p-primary-action]=\"primaryAction\" [p-title]=\"title\">\n  <div class=\"po-row\">\n    <po-accordion class=\"po-md-12 po-mt-1 po-mb-1\">\n      <po-accordion-item *ngFor=\"let detail of details\" [p-label]=\"formatDetailItemTitle(detail)\">\n        <div *ngIf=\"detail.type\" class=\"po-row po-mb-1\">\n          <po-tag [p-color]=\"typeColor(detail.type)\" [p-value]=\"typeValue(detail.type)\"></po-tag>\n        </div>\n\n        <div class=\"po-row\">\n          <p>{{ detail.detailedMessage }}</p>\n        </div>\n      </po-accordion-item>\n    </po-accordion>\n  </div>\n</po-modal>\n"
        })
    ], PoHttpInterceptorDetailComponent);
    return PoHttpInterceptorDetailComponent;
}());
export { PoHttpInterceptorDetailComponent };
