// @po-ui/ng-components/esm2015/lib/components/po-disclaimer/po-disclaimer-base.component.js
import { __decorate, __metadata } from "tslib";
import { EventEmitter, Input, Output, Directive } from '@angular/core';
import { convertToBoolean } from '../../utils/util';
const PO_DISCLAIMER_TYPES = ['default', 'danger'];
const PO_DISCLAIMER_DEFAULT_TYPE = 'default';
/**
 * @docsPrivate
 *
 * @description
 *
 * O componente po-disclaimer é responsável por representar tags.
 * Seu uso é recomendado em buscas e em campos onde é necessário representar objetos selecionados,
 * como por exemplo, no po-multi-select.
 *
 */
let PoDisclaimerBaseComponent = class PoDisclaimerBaseComponent {
    constructor() {
        this._type = 'default';
        this._hideClose = false;
        this.showDisclaimer = true;
        /**
         * Evento disparado ao fechar o disclaimer.
         * Para este evento será passado como parâmetro um objeto com value, label e property.
         */
        this.closeAction = new EventEmitter();
    }
    /**
     * @description
     *
     * Esta propriedade esconde o botão para fechamento do po-disclaimer, ao utilizar esta propriedade
     * sem passar valor a mesma é setada como false, onde o botão de fechamento está visível.
     *
     * @default false
     */
    set hideClose(value) {
        this._hideClose = value === '' ? true : convertToBoolean(value);
    }
    get hideClose() {
        return this._hideClose;
    }
    /**
     * @description
     *
     * Tipo do po-disclaimer. Pode ser 'default' ou 'danger'.
     *
     * @default default
     * @optional
     */
    set type(type) {
        this._type = PO_DISCLAIMER_TYPES.includes(type) ? type : PO_DISCLAIMER_DEFAULT_TYPE;
    }
    get type() {
        return this._type;
    }
    close() {
        this.showDisclaimer = false;
        this.closeAction.emit({ value: this.value, label: this.label, property: this.property });
    }
    getLabel() {
        return this.label ? this.label : this.value;
    }
};
__decorate([
    Input('p-label'),
    __metadata("design:type", String)
], PoDisclaimerBaseComponent.prototype, "label", void 0);
__decorate([
    Input('p-value'),
    __metadata("design:type", String)
], PoDisclaimerBaseComponent.prototype, "value", void 0);
__decorate([
    Input('p-property'),
    __metadata("design:type", String)
], PoDisclaimerBaseComponent.prototype, "property", void 0);
__decorate([
    Input('p-hide-close'),
    __metadata("design:type", Boolean),
    __metadata("design:paramtypes", [Boolean])
], PoDisclaimerBaseComponent.prototype, "hideClose", null);
__decorate([
    Output('p-close-action'),
    __metadata("design:type", EventEmitter)
], PoDisclaimerBaseComponent.prototype, "closeAction", void 0);
__decorate([
    Input('p-type'),
    __metadata("design:type", String),
    __metadata("design:paramtypes", [String])
], PoDisclaimerBaseComponent.prototype, "type", null);
PoDisclaimerBaseComponent = __decorate([
    Directive()
], PoDisclaimerBaseComponent);
export { PoDisclaimerBaseComponent };
