// @po-ui/ng-components/esm5/lib/components/po-field/po-lookup/po-lookup-modal/po-lookup-modal.component.js
import { __decorate, __extends, __metadata } from "tslib";
import { Component, ElementRef, ViewChild } from '@angular/core';
import { fromEvent } from 'rxjs';
import { debounceTime, filter } from 'rxjs/operators';
import { PoLookupModalBaseComponent } from '../po-lookup-modal/po-lookup-modal-base.component';
/**
 * @docsPrivate
 *
 * @docsExtends PoLookupModalBaseComponent
 */
var PoLookupModalComponent = /** @class */ (function (_super) {
    __extends(PoLookupModalComponent, _super);
    function PoLookupModalComponent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.keyUpObservable = null;
        _this.containerHeight = 375;
        _this.tableHeight = 370;
        return _this;
    }
    PoLookupModalComponent.prototype.ngOnInit = function () {
        _super.prototype.ngOnInit.call(this);
        this.initializeEventInput();
        this.setTableHeight();
    };
    PoLookupModalComponent.prototype.initializeEventInput = function () {
        var _this = this;
        this.keyUpObservable = fromEvent(this.inputSearchEl.nativeElement, 'keyup').pipe(filter(function (e) { return _this.validateEnterPressed(e); }), debounceTime(400));
        this.keyUpObservable.subscribe(function () {
            _this.search();
        });
    };
    PoLookupModalComponent.prototype.openModal = function () {
        this.poModal.open();
    };
    PoLookupModalComponent.prototype.sortBy = function (sort) {
        this.sort = sort;
    };
    PoLookupModalComponent.prototype.setTableHeight = function () {
        if (window.innerHeight < 615) {
            this.tableHeight -= 50;
            this.containerHeight -= 50;
        }
    };
    PoLookupModalComponent.prototype.validateEnterPressed = function (e) {
        return e.keyCode === 13;
    };
    __decorate([
        ViewChild('inpsearch', { static: true }),
        __metadata("design:type", ElementRef)
    ], PoLookupModalComponent.prototype, "inputSearchEl", void 0);
    PoLookupModalComponent = __decorate([
        Component({
            selector: 'po-lookup-modal',
            template: "<po-modal\n  p-click-out=\"false\"\n  p-hide-close=\"false\"\n  p-size=\"lg\"\n  [p-primary-action]=\"primaryAction\"\n  [p-secondary-action]=\"secondaryAction\"\n  [p-title]=\"title\"\n>\n  <po-field-container class=\"po-lookup-header po-md-6 po-pull-right\" [p-optional]=\"false\">\n    <div class=\"po-field-container-content\">\n      <input\n        #inpsearch\n        class=\"po-input po-input-icon-right\"\n        name=\"contentSearch\"\n        [(ngModel)]=\"searchValue\"\n        [placeholder]=\"literals.modalPlaceholder\"\n        type=\"text\"\n      />\n\n      <div class=\"po-field-icon-container-right\">\n        <span #iconLookup class=\"po-icon po-field-icon po-icon-search\" (click)=\"search()\"> </span>\n      </div>\n    </div>\n  </po-field-container>\n\n  <div class=\"po-row po-lookup-container-table\" [style.height.px]=\"containerHeight\">\n    <po-table\n      class=\"po-md-12\"\n      p-checkbox=\"true\"\n      p-hide-detail=\"true\"\n      p-single-select=\"true\"\n      p-sort=\"true\"\n      [p-columns]=\"columns\"\n      [p-height]=\"tableHeight\"\n      [p-items]=\"items\"\n      [p-literals]=\"tableLiterals\"\n      [p-loading]=\"isLoading\"\n      [p-show-more-disabled]=\"!hasNext\"\n      (p-show-more)=\"showMoreEvent()\"\n      (p-sort-by)=\"sortBy($event)\"\n    >\n    </po-table>\n  </div>\n</po-modal>\n"
        })
    ], PoLookupModalComponent);
    return PoLookupModalComponent;
}(PoLookupModalBaseComponent));
export { PoLookupModalComponent };
