// @po-ui/ng-components/esm5/lib/components/po-accordion/po-accordion-item-header/po-accordion-item-header.component.js
import { __decorate, __metadata } from "tslib";
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output } from '@angular/core';
var PoAccordionItemHeaderComponent = /** @class */ (function () {
    function PoAccordionItemHeaderComponent() {
        this.expanded = false;
        this.toggle = new EventEmitter();
    }
    PoAccordionItemHeaderComponent.prototype.onClick = function () {
        this.expanded = !this.expanded;
        this.toggle.emit(this.expanded);
    };
    __decorate([
        Input('p-expanded'),
        __metadata("design:type", Boolean)
    ], PoAccordionItemHeaderComponent.prototype, "expanded", void 0);
    __decorate([
        Input('p-label'),
        __metadata("design:type", String)
    ], PoAccordionItemHeaderComponent.prototype, "label", void 0);
    __decorate([
        Output('p-toggle'),
        __metadata("design:type", Object)
    ], PoAccordionItemHeaderComponent.prototype, "toggle", void 0);
    PoAccordionItemHeaderComponent = __decorate([
        Component({
            selector: 'po-accordion-item-header',
            template: "<header class=\"po-accordion-item-header\">\n  <button class=\"po-accordion-item-header-button po-clickable\" (click)=\"onClick()\">\n    <div class=\"po-text-ellipsis po-accordion-item-header-title\">{{ label }}</div>\n    <span #icon class=\"po-icon po-accordion-item-header-icon po-icon-arrow-down\"> </span>\n  </button>\n</header>\n",
            changeDetection: ChangeDetectionStrategy.OnPush
        })
    ], PoAccordionItemHeaderComponent);
    return PoAccordionItemHeaderComponent;
}());
export { PoAccordionItemHeaderComponent };
