// @po-ui/ng-components/esm2015/lib/components/po-accordion/po-accordion-item/po-accordion-item.component.js
import { __decorate, __metadata } from "tslib";
import { Component, Input, TemplateRef, ViewChild } from '@angular/core';
import { PoAccordionService } from '../services/po-accordion.service';
/**
 * @description
 *
 * Componente utilizado para renderizar os itens do `po-accordion`.
 *
 * O componente `po-accordion` já faz o controle de abertura e fechamento dos itens automaticamente,
 * mas caso houver a necessidade de abrir algum dos `po-accordion-item` via Typescript, pode ser feita da seguinte forma:
 *
 * ```
 * <po-accordion>
 *   <po-accordion-item p-label="PO Accordion 1" #item1>
 *      Accordion 1
 *   </po-accordion-item>
 *
 *   <po-accordion-item p-label="PO Accordion 2">
 *      Accordion 2
 *   </po-accordion-item>
 * </po-accordion>
 * ```
 *
 * e no typescript pode-se utilizar o `@ViewChild`:
 *
 * ```
 *  @ViewChild(PoAccordionItemComponent, { static: true }) item1: PoAccordionItemComponent;
 *
 *  ngAfterContentInit() {
 *    // ou utilizar o método collapse()
 *    this.item1.expand();
 *  }
 * ```
 */
let PoAccordionItemComponent = class PoAccordionItemComponent {
    constructor(accordionService) {
        this.accordionService = accordionService;
    }
    /**
     * Método para colapsar o `po-accordion-item`.
     */
    collapse() {
        this.expanded = false;
        this.accordionService.sendToParentAccordionItemClicked(this);
    }
    /**
     * Método para expandir o `po-accordion-item`.
     */
    expand() {
        this.expanded = true;
        this.accordionService.sendToParentAccordionItemClicked(this);
    }
};
PoAccordionItemComponent.ctorParameters = () => [
    { type: PoAccordionService }
];
__decorate([
    Input('p-label'),
    __metadata("design:type", String)
], PoAccordionItemComponent.prototype, "label", void 0);
__decorate([
    ViewChild(TemplateRef, { static: true }),
    __metadata("design:type", TemplateRef)
], PoAccordionItemComponent.prototype, "templateRef", void 0);
PoAccordionItemComponent = __decorate([
    Component({
        selector: 'po-accordion-item',
        template: "<ng-template>\n  <ng-content></ng-content>\n</ng-template>\n"
    }),
    __metadata("design:paramtypes", [PoAccordionService])
], PoAccordionItemComponent);
export { PoAccordionItemComponent };
