// @po-ui/ng-components/esm5/lib/components/po-tree-view/po-tree-view-item-header/po-tree-view-item-header.component.js
import { __decorate, __metadata } from "tslib";
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output, ViewChild } from '@angular/core';
var PoTreeViewItemHeaderComponent = /** @class */ (function () {
    function PoTreeViewItemHeaderComponent() {
        this.selectable = false;
        this.expanded = new EventEmitter();
        this.selected = new EventEmitter();
    }
    Object.defineProperty(PoTreeViewItemHeaderComponent.prototype, "hasSubItems", {
        get: function () {
            return !!(this.item.subItems && this.item.subItems.length);
        },
        enumerable: true,
        configurable: true
    });
    __decorate([
        ViewChild('inputCheckbox'),
        __metadata("design:type", Object)
    ], PoTreeViewItemHeaderComponent.prototype, "inputCheckbox", void 0);
    __decorate([
        Input('p-item'),
        __metadata("design:type", Object)
    ], PoTreeViewItemHeaderComponent.prototype, "item", void 0);
    __decorate([
        Input('p-selectable'),
        __metadata("design:type", Boolean)
    ], PoTreeViewItemHeaderComponent.prototype, "selectable", void 0);
    __decorate([
        Output('p-expanded'),
        __metadata("design:type", Object)
    ], PoTreeViewItemHeaderComponent.prototype, "expanded", void 0);
    __decorate([
        Output('p-selected'),
        __metadata("design:type", Object)
    ], PoTreeViewItemHeaderComponent.prototype, "selected", void 0);
    PoTreeViewItemHeaderComponent = __decorate([
        Component({
            selector: 'po-tree-view-item-header',
            template: "<div class=\"po-tree-view-item-header\">\n  <button *ngIf=\"hasSubItems\" class=\"po-tree-view-item-header-button\" (click)=\"expanded.emit($event)\">\n    <span\n      class=\"po-icon po-icon-arrow-down po-tree-view-item-header-button-icon\"\n      [class.po-tree-view-item-header-button-icon-transform]=\"item.expanded\"\n    >\n    </span>\n  </button>\n\n  <ng-container *ngIf=\"selectable; then checkboxTemplate; else labelTemplate\"></ng-container>\n</div>\n\n<ng-template #labelTemplate>\n  <span class=\"po-tree-view-item-header-label\" [class.po-tree-view-item-header-padding]=\"!hasSubItems\">\n    {{ item.label }}\n  </span>\n</ng-template>\n\n<ng-template #checkboxTemplate>\n  <po-checkbox\n    class=\"po-tree-view-item-header-checkbox\"\n    [class.po-tree-view-item-header-padding]=\"!hasSubItems\"\n    [p-label]=\"item.label\"\n    [(ngModel)]=\"item.selected\"\n    (p-change)=\"selected.emit(item)\"\n  >\n  </po-checkbox>\n</ng-template>\n",
            changeDetection: ChangeDetectionStrategy.OnPush
        })
    ], PoTreeViewItemHeaderComponent);
    return PoTreeViewItemHeaderComponent;
}());
export { PoTreeViewItemHeaderComponent };
