// @po-ui/ng-components/esm5/lib/components/po-tag/po-tag.component.js
import { __decorate, __extends } from "tslib";
import { ChangeDetectionStrategy, Component } from '@angular/core';
import { PoTagBaseComponent } from './po-tag-base.component';
import { PoTagIcon } from './enums/po-tag-icon.enum';
import { PoTagType } from './enums/po-tag-type.enum';
var poTagTypeDefault = 'po-tag-' + PoTagType.Info;
/**
 * @docsExtends PoTagBaseComponent
 *
 * @example
 *
 * <example name="po-tag-basic" title="PO Tag Basic">
 *  <file name="sample-po-tag-basic/sample-po-tag-basic.component.html"> </file>
 *  <file name="sample-po-tag-basic/sample-po-tag-basic.component.ts"> </file>
 * </example>
 *
 * <example name="po-tag-labs" title="PO Tag Labs">
 *  <file name="sample-po-tag-labs/sample-po-tag-labs.component.html"> </file>
 *  <file name="sample-po-tag-labs/sample-po-tag-labs.component.ts"> </file>
 * </example>
 *
 * <example name="po-tag-bank-account" title="PO Tag - Bank Account">
 *  <file name="sample-po-tag-bank-account/sample-po-tag-bank-account.component.html"> </file>
 *  <file name="sample-po-tag-bank-account/sample-po-tag-bank-account.component.ts"> </file>
 * </example>
 */
var PoTagComponent = /** @class */ (function (_super) {
    __extends(PoTagComponent, _super);
    function PoTagComponent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    PoTagComponent.prototype.ngOnInit = function () {
        this.isClickable = this.click.observers.length > 0;
    };
    Object.defineProperty(PoTagComponent.prototype, "iconFromType", {
        get: function () {
            switch (this.type) {
                case PoTagType.Danger:
                    return PoTagIcon.Danger;
                case PoTagType.Info:
                    return PoTagIcon.Info;
                case PoTagType.Success:
                    return PoTagIcon.Success;
                case PoTagType.Warning:
                    return PoTagIcon.Warning;
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoTagComponent.prototype, "iconTypeString", {
        get: function () {
            return typeof this.icon === 'string';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoTagComponent.prototype, "tagColor", {
        get: function () {
            if (this.type) {
                return this.inverse ? "po-tag-" + this.type + "-inverse" : "po-tag-" + this.type;
            }
            if (this.color) {
                return this.inverse ? "po-text-" + this.color : "po-" + this.color;
            }
            return this.inverse ? poTagTypeDefault + "-inverse" : poTagTypeDefault;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoTagComponent.prototype, "tagOrientation", {
        get: function () {
            return this.orientation === this.poTagOrientation.Horizontal;
        },
        enumerable: true,
        configurable: true
    });
    PoTagComponent.prototype.onClick = function () {
        var submittedTagItem = { value: this.value, type: this.type };
        this.click.emit(submittedTagItem);
    };
    PoTagComponent.prototype.onKeyPressed = function (event) {
        event.preventDefault();
        event.stopPropagation();
        this.onClick();
    };
    PoTagComponent = __decorate([
        Component({
            selector: 'po-tag',
            template: "<div class=\"po-tag-container\" [class.po-tag-container-horizontal]=\"tagOrientation\">\n  <div *ngIf=\"label\" class=\"po-tag-title po-text-nowrap\">\n    <span class=\"po-tag-label\">{{ tagOrientation ? label + ':' : label }}</span>\n  </div>\n\n  <div class=\"po-tag-sub-container\">\n    <div\n      class=\"po-tag\"\n      [class.po-clickable]=\"isClickable\"\n      [class.po-tag-inverse]=\"inverse\"\n      [ngClass]=\"tagColor\"\n      tabindex=\"0\"\n      (click)=\"onClick()\"\n      (keydown.enter)=\"onKeyPressed($event)\"\n      (keydown.space)=\"$event.preventDefault()\"\n      (keyup.space)=\"onKeyPressed($event)\"\n    >\n      <span *ngIf=\"icon\" class=\"po-icon\" [ngClass]=\"!type && iconTypeString ? icon : iconFromType\"></span>\n      <span class=\"po-tag-value\">{{ value }}</span>\n    </div>\n  </div>\n</div>\n",
            changeDetection: ChangeDetectionStrategy.OnPush
        })
    ], PoTagComponent);
    return PoTagComponent;
}(PoTagBaseComponent));
export { PoTagComponent };
