// @po-ui/ng-components/esm5/lib/components/po-field/po-multiselect/po-multiselect-item/po-multiselect-item.component.js
import { __decorate, __metadata } from "tslib";
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output } from '@angular/core';
/**
 * @docsPrivate
 *
 * @description
 *
 * Componente que constrói cada item no dropdown, contendo o checkbox e o label.
 */
var PoMultiselectItemComponent = /** @class */ (function () {
    function PoMultiselectItemComponent() {
        /** Esta propriedade indica se o campo está selecionado ou não. */
        this.selected = false;
        /** Evento que será disparado toda vez que o usuário marcar ou desmarcar um item. */
        this.change = new EventEmitter();
    }
    PoMultiselectItemComponent.prototype.itemClicked = function () {
        this.selected = !this.selected;
        this.change.emit(this.selected);
    };
    __decorate([
        Input('p-label'),
        __metadata("design:type", String)
    ], PoMultiselectItemComponent.prototype, "label", void 0);
    __decorate([
        Input('p-selected'),
        __metadata("design:type", Boolean)
    ], PoMultiselectItemComponent.prototype, "selected", void 0);
    __decorate([
        Output('p-change'),
        __metadata("design:type", Object)
    ], PoMultiselectItemComponent.prototype, "change", void 0);
    PoMultiselectItemComponent = __decorate([
        Component({
            selector: 'po-multiselect-item',
            template: "<li [class.po-multiselect-item-selected]=\"selected\" (click)=\"itemClicked()\">\n  <a class=\"po-multiselect-item\">\n    <input\n      [checked]=\"selected\"\n      class=\"po-multiselect-checkbox-input\"\n      [class.po-multiselect-checkbox-input-checked]=\"selected\"\n      type=\"checkbox\"\n    />\n\n    <label class=\"po-multiselect-checkbox-label po-clickable\">\n      {{ label }}\n    </label>\n  </a>\n</li>\n",
            changeDetection: ChangeDetectionStrategy.OnPush
        })
    ], PoMultiselectItemComponent);
    return PoMultiselectItemComponent;
}());
export { PoMultiselectItemComponent };
