// @po-ui/ng-components/esm5/lib/components/po-chart/po-chart-types/po-chart-gauge/po-chart-gauge.component.js
import { __assign, __decorate, __extends, __metadata, __read } from "tslib";
import { AfterViewInit, ChangeDetectorRef, Component, ComponentFactoryResolver, ElementRef, NgZone, Renderer2, ViewChild, ViewContainerRef } from '@angular/core';
import { PoChartCircular } from '../po-chart-circular/po-chart-circular';
import { poChartGaugeStartAngle } from '../po-chart-circular/po-chart-circular.constant';
import { PoChartGaugeTextContentComponent } from './po-chart-gauge-text-content/po-chart-gauge-text-content.component';
var PoChartGaugeComponent = /** @class */ (function (_super) {
    __extends(PoChartGaugeComponent, _super);
    /* istanbul ignore next */
    function PoChartGaugeComponent(changeDetection, componentFactoryResolver, el, ngZone, renderer) {
        var _this = _super.call(this, el, ngZone, renderer) || this;
        _this.changeDetection = changeDetection;
        _this.componentFactoryResolver = componentFactoryResolver;
        _this.chartItemStartAngle = poChartGaugeStartAngle;
        _this._series = [];
        return _this;
    }
    Object.defineProperty(PoChartGaugeComponent.prototype, "series", {
        get: function () {
            return this._series;
        },
        set: function (series) {
            this._series = this.getGaugeSerie(series);
        },
        enumerable: true,
        configurable: true
    });
    PoChartGaugeComponent.prototype.ngAfterViewInit = function () {
        this.drawBasePath();
        this.createComponent();
    };
    PoChartGaugeComponent.prototype.createComponent = function () {
        var factory = this.componentFactoryResolver.resolveComponentFactory(PoChartGaugeTextContentComponent);
        var componentRef = this.svgContainerRef.createComponent(factory);
        var instance = componentRef.instance;
        instance.serie = this.series.length && this.series[0];
        instance.gaugeWidth = this.getGaugeBaseWidth();
        this.resizeListenerSubscription(instance);
    };
    PoChartGaugeComponent.prototype.checkGaugeValueLimits = function (value) {
        if (value <= 0) {
            return 0;
        }
        else if (value >= 100) {
            return 100;
        }
        else {
            return value;
        }
    };
    PoChartGaugeComponent.prototype.drawBasePath = function () {
        var basePath = this.el.nativeElement.querySelector('.po-chart-gauge-base-path');
        this.drawPath(basePath, this.chartItemStartAngle, 0);
    };
    PoChartGaugeComponent.prototype.getGaugeBaseWidth = function () {
        var basePath = this.el.nativeElement.querySelector('.po-chart-gauge-base-path');
        if (basePath) {
            return basePath.getBoundingClientRect().width;
        }
    };
    PoChartGaugeComponent.prototype.getGaugeSerie = function (series) {
        if (series === void 0) { series = []; }
        var _a = __read(series, 1), serie = _a[0];
        if (serie && typeof serie === 'object') {
            return [
                __assign(__assign({}, serie), { color: this.colors[0], value: this.checkGaugeValueLimits(serie.value) })
            ];
        }
        return [];
    };
    PoChartGaugeComponent.prototype.resizeListenerSubscription = function (instance) {
        var _this = this;
        this.windowResizeEmitter.subscribe(function () {
            instance.gaugeWidth = _this.getGaugeBaseWidth();
            _this.changeDetection.detectChanges();
        });
    };
    PoChartGaugeComponent.ctorParameters = function () { return [
        { type: ChangeDetectorRef },
        { type: ComponentFactoryResolver },
        { type: ElementRef },
        { type: NgZone },
        { type: Renderer2 }
    ]; };
    __decorate([
        ViewChild('svgContainer', { static: true, read: ViewContainerRef }),
        __metadata("design:type", ViewContainerRef)
    ], PoChartGaugeComponent.prototype, "svgContainerRef", void 0);
    PoChartGaugeComponent = __decorate([
        Component({
            selector: 'po-chart-gauge',
            template: "<div #chartBody class=\"po-chart-body\">\n  <div #svgContainer class=\"po-chart-svg-container\"></div>\n\n  <div #tooltipElement *ngIf=\"!isChartGaugeType\" class=\"po-chart-tooltip po-tooltip po-invisible\">\n    <div class=\"po-tooltip-arrow po-arrow-bottom\"></div>\n    <div class=\"po-tooltip-content\">{{ tooltipText }}</div>\n  </div>\n</div>\n"
        }),
        __metadata("design:paramtypes", [ChangeDetectorRef,
            ComponentFactoryResolver,
            ElementRef,
            NgZone,
            Renderer2])
    ], PoChartGaugeComponent);
    return PoChartGaugeComponent;
}(PoChartCircular));
export { PoChartGaugeComponent };
