// @po-ui/ng-components/esm5/lib/components/po-field/po-upload/po-upload-drag-drop/po-upload-drag-drop.component.js
import { __decorate, __metadata } from "tslib";
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, ElementRef, EventEmitter, Input, Output, ViewChild } from '@angular/core';
import { convertToInt } from '../../../../utils/util';
import { PoUploadDragDropAreaComponent } from './po-upload-drag-drop-area/po-upload-drag-drop-area.component';
var PoUploadDragDropHeightDefault = 320;
var PoUploadDragDropHeightMin = 160;
var PoUploadDragDropComponent = /** @class */ (function () {
    function PoUploadDragDropComponent(changeDetector) {
        this.changeDetector = changeDetector;
        this._dragDropHeight = PoUploadDragDropHeightDefault;
        this.isDragOver = false;
        this.fileChange = new EventEmitter();
        this.selectFiles = new EventEmitter();
    }
    Object.defineProperty(PoUploadDragDropComponent.prototype, "dragDropHeight", {
        get: function () {
            return this._dragDropHeight;
        },
        set: function (value) {
            var dragDropHeight = convertToInt(value, PoUploadDragDropHeightDefault);
            this._dragDropHeight = dragDropHeight < PoUploadDragDropHeightMin ? PoUploadDragDropHeightMin : dragDropHeight;
        },
        enumerable: true,
        configurable: true
    });
    PoUploadDragDropComponent.prototype.focus = function () {
        this.dragDropAreaComponent.focus();
    };
    PoUploadDragDropComponent.prototype.onAreaElement = function (element) {
        this.areaElement = element;
        // necessário para não ocorrer o erro HasBeenChecked
        this.changeDetector.detectChanges();
    };
    PoUploadDragDropComponent.prototype.onDragLeave = function () {
        this.isDragOver = false;
    };
    PoUploadDragDropComponent.prototype.onDragOver = function () {
        this.isDragOver = true;
    };
    PoUploadDragDropComponent.prototype.onFileChange = function (files) {
        this.isDragOver = false;
        this.fileChange.emit(files);
    };
    PoUploadDragDropComponent.ctorParameters = function () { return [
        { type: ChangeDetectorRef }
    ]; };
    __decorate([
        ViewChild('dragDropOverlay', { read: ElementRef }),
        __metadata("design:type", ElementRef)
    ], PoUploadDragDropComponent.prototype, "dragDropOverlayElement", void 0);
    __decorate([
        ViewChild(PoUploadDragDropAreaComponent, { static: true }),
        __metadata("design:type", PoUploadDragDropAreaComponent)
    ], PoUploadDragDropComponent.prototype, "dragDropAreaComponent", void 0);
    __decorate([
        Input('p-directory-compatible'),
        __metadata("design:type", Boolean)
    ], PoUploadDragDropComponent.prototype, "directoryCompatible", void 0);
    __decorate([
        Input('p-disabled'),
        __metadata("design:type", Boolean)
    ], PoUploadDragDropComponent.prototype, "disabled", void 0);
    __decorate([
        Input('p-drag-drop-height'),
        __metadata("design:type", Number),
        __metadata("design:paramtypes", [Number])
    ], PoUploadDragDropComponent.prototype, "dragDropHeight", null);
    __decorate([
        Input('p-literals'),
        __metadata("design:type", Object)
    ], PoUploadDragDropComponent.prototype, "literals", void 0);
    __decorate([
        Output('p-file-change'),
        __metadata("design:type", EventEmitter)
    ], PoUploadDragDropComponent.prototype, "fileChange", void 0);
    __decorate([
        Output('p-select-files'),
        __metadata("design:type", EventEmitter)
    ], PoUploadDragDropComponent.prototype, "selectFiles", void 0);
    PoUploadDragDropComponent = __decorate([
        Component({
            selector: 'po-upload-drag-drop',
            template: "<po-upload-drag-drop-area-overlay\n  #dragDropOverlay\n  *ngIf=\"isDragOver\"\n  [p-directory-compatible]=\"directoryCompatible\"\n  [p-disabled]=\"disabled\"\n  [p-literals]=\"literals\"\n  [p-target]=\"dragDropAreaComponent.elementRef\"\n  (p-area-element)=\"onAreaElement($event)\"\n>\n</po-upload-drag-drop-area-overlay>\n\n<po-upload-drag-drop-area\n  p-upload-drag-drop\n  [p-area-element]=\"areaElement\"\n  [p-directory-compatible]=\"directoryCompatible\"\n  [p-disabled]=\"disabled\"\n  [p-height]=\"dragDropHeight\"\n  [p-literals]=\"literals\"\n  [p-overlay]=\"false\"\n  (p-drag-leave)=\"onDragLeave()\"\n  (p-drag-over)=\"onDragOver()\"\n  (p-file-change)=\"onFileChange($event)\"\n  (p-select-files)=\"selectFiles.emit()\"\n>\n</po-upload-drag-drop-area>\n",
            changeDetection: ChangeDetectionStrategy.OnPush
        }),
        __metadata("design:paramtypes", [ChangeDetectorRef])
    ], PoUploadDragDropComponent);
    return PoUploadDragDropComponent;
}());
export { PoUploadDragDropComponent };
