// @po-ui/ng-components/esm5/lib/components/po-menu-panel/po-menu-panel-item/po-menu-panel-item.component.js
import { __decorate, __metadata } from "tslib";
import { Component, Input } from '@angular/core';
import { PoMenuPanelItemsService } from '../services/po-menu-panel-items.service';
/**
 * @docsPrivate
 *
 * @description
 *
 * Componente que implementa cada item do po-menu-panel.
 */
var PoMenuPanelItemComponent = /** @class */ (function () {
    function PoMenuPanelItemComponent(menuItemsService) {
        this.menuItemsService = menuItemsService;
    }
    PoMenuPanelItemComponent.prototype.ngOnDestroy = function () {
        this.itemsSubscription.unsubscribe();
    };
    PoMenuPanelItemComponent.prototype.ngOnInit = function () {
        // subscribe to menu component messages
        this.subscribeMenuClickedFromParent();
    };
    PoMenuPanelItemComponent.prototype.clickMenuItem = function (event) {
        if (!(event.ctrlKey || event.metaKey)) {
            event.preventDefault();
            // Emmit to parent
            this.menuItemsService.sendToParentMenuClicked(this.menuItemInternal);
        }
    };
    PoMenuPanelItemComponent.prototype.activateMenu = function (menu) {
        this.menuItemInternal.isSelected = this.menuItemInternal.id === menu.id;
    };
    PoMenuPanelItemComponent.prototype.processMenuItem = function (menu) {
        if (this.menuItemInternal.type === 'internalLink') {
            this.activateMenu(menu.active);
        }
    };
    PoMenuPanelItemComponent.prototype.subscribeMenuClickedFromParent = function () {
        var _this = this;
        this.itemsSubscription = this.menuItemsService.receiveFromParentMenuClicked().subscribe(function (menu) {
            _this.processMenuItem(menu);
        });
    };
    PoMenuPanelItemComponent.ctorParameters = function () { return [
        { type: PoMenuPanelItemsService }
    ]; };
    __decorate([
        Input('p-menu-item-internal'),
        __metadata("design:type", Object)
    ], PoMenuPanelItemComponent.prototype, "menuItemInternal", void 0);
    PoMenuPanelItemComponent = __decorate([
        Component({
            selector: 'po-menu-panel-item',
            template: "<!-- menu com link interno -->\n<a *ngIf=\"menuItemInternal.type === 'internalLink'\" [routerLink]=\"menuItemInternal.link\">\n  <ng-container *ngTemplateOutlet=\"menuItemTemplate\"></ng-container>\n</a>\n\n<!-- menu com link externo -->\n<a *ngIf=\"menuItemInternal.type === 'externalLink'\" [href]=\"menuItemInternal.link\">\n  <ng-container *ngTemplateOutlet=\"menuItemTemplate\"></ng-container>\n</a>\n\n<!-- menu sem link -->\n<a *ngIf=\"menuItemInternal.type === 'noLink'\" href=\"javascript:;\">\n  <ng-container *ngTemplateOutlet=\"menuItemTemplate\"></ng-container>\n</a>\n\n<ng-template #menuItemTemplate>\n  <div\n    class=\"po-menu-panel-item\"\n    p-tooltip-position=\"right\"\n    [class.po-menu-panel-item-selected]=\"menuItemInternal.isSelected\"\n    [p-tooltip]=\"menuItemInternal.label\"\n    (click)=\"clickMenuItem($event)\"\n  >\n    <span class=\"po-icon {{ menuItemInternal.icon }}\"></span>\n  </div>\n</ng-template>\n"
        }),
        __metadata("design:paramtypes", [PoMenuPanelItemsService])
    ], PoMenuPanelItemComponent);
    return PoMenuPanelItemComponent;
}());
export { PoMenuPanelItemComponent };
