// @po-ui/ng-components/esm5/lib/components/po-field/po-upload/po-upload-drag-drop/po-upload-drag-drop-area/po-upload-drag-drop-area.component.js
import { __decorate, __metadata } from "tslib";
import { ChangeDetectionStrategy, Component, ElementRef, EventEmitter, Input, Output, ViewChild } from '@angular/core';
var PoUploadDragDropAreaComponent = /** @class */ (function () {
    function PoUploadDragDropAreaComponent(elementRef) {
        this.elementRef = elementRef;
        this.selectFiles = new EventEmitter();
    }
    PoUploadDragDropAreaComponent.prototype.focus = function () {
        this.selectFilesLinkElement.nativeElement.focus();
    };
    PoUploadDragDropAreaComponent.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    __decorate([
        Input('p-directory-compatible'),
        __metadata("design:type", Boolean)
    ], PoUploadDragDropAreaComponent.prototype, "directoryCompatible", void 0);
    __decorate([
        Input('p-disabled'),
        __metadata("design:type", Boolean)
    ], PoUploadDragDropAreaComponent.prototype, "disabled", void 0);
    __decorate([
        Input('p-height'),
        __metadata("design:type", Number)
    ], PoUploadDragDropAreaComponent.prototype, "height", void 0);
    __decorate([
        Input('p-literals'),
        __metadata("design:type", Object)
    ], PoUploadDragDropAreaComponent.prototype, "literals", void 0);
    __decorate([
        Input('p-overlay'),
        __metadata("design:type", Boolean)
    ], PoUploadDragDropAreaComponent.prototype, "overlay", void 0);
    __decorate([
        Output('p-select-files'),
        __metadata("design:type", EventEmitter)
    ], PoUploadDragDropAreaComponent.prototype, "selectFiles", void 0);
    __decorate([
        ViewChild('selectFilesLink'),
        __metadata("design:type", ElementRef)
    ], PoUploadDragDropAreaComponent.prototype, "selectFilesLinkElement", void 0);
    PoUploadDragDropAreaComponent = __decorate([
        Component({
            selector: 'po-upload-drag-drop-area',
            template: "<div\n  class=\"po-upload-drag-drop-area\"\n  [ngClass]=\"{ 'po-upload-drag-drop-area-disabled': disabled }\"\n  [style.height.px]=\"height\"\n>\n  <div class=\"po-upload-drag-drop-area-container\">\n    <ng-container *ngIf=\"overlay && !disabled; then overlayTemplate; else defaultTemplate\"> </ng-container>\n  </div>\n</div>\n\n<ng-template #overlayTemplate>\n  <span class=\"po-upload-drag-drop-area-overlay-icon po-icon po-icon-upload-cloud\"></span>\n\n  <div class=\"po-upload-drag-drop-area-overlay-label\">\n    {{ directoryCompatible ? literals?.dropFoldersHere : literals?.dropFilesHere }}\n  </div>\n</ng-template>\n\n<ng-template #defaultTemplate>\n  <span class=\"po-upload-drag-drop-area-icon po-icon po-icon-upload-cloud\"></span>\n\n  <div class=\"po-upload-drag-drop-area-label\">\n    {{ directoryCompatible ? literals?.dragFoldersHere : literals?.dragFilesHere }}\n  </div>\n\n  <button\n    #selectFilesLink\n    class=\"po-upload-drag-drop-area-select-files\"\n    [disabled]=\"disabled\"\n    [ngClass]=\"{ 'po-clickable': !disabled }\"\n    (click)=\"selectFiles.emit()\"\n  >\n    {{ directoryCompatible ? literals?.selectFolderOnComputer : literals?.selectFilesOnComputer }}\n  </button>\n</ng-template>\n",
            changeDetection: ChangeDetectionStrategy.OnPush
        }),
        __metadata("design:paramtypes", [ElementRef])
    ], PoUploadDragDropAreaComponent);
    return PoUploadDragDropAreaComponent;
}());
export { PoUploadDragDropAreaComponent };
