// @po-ui/ng-components/esm5/lib/components/po-chart/po-chart-types/po-chart-dynamic-type.component.js
import { __decorate, __metadata } from "tslib";
import { ElementRef, ViewChild, Directive } from '@angular/core';
import { Subject } from 'rxjs';
import { PoChartType } from '../enums/po-chart-type.enum';
var Padding = 24;
/* tslint:disable:directive-class-suffix */
var PoChartDynamicTypeComponent = /** @class */ (function () {
    function PoChartDynamicTypeComponent() {
        this.innerRadius = 0;
        this.onSerieClick = new Subject();
        this.onSerieHover = new Subject();
        this.series = [];
    }
    Object.defineProperty(PoChartDynamicTypeComponent.prototype, "isChartGaugeType", {
        get: function () {
            return this.type === PoChartType.Gauge;
        },
        enumerable: true,
        configurable: true
    });
    PoChartDynamicTypeComponent.prototype.calculateSVGContainerDimensions = function (chartWrapperElement, chartHeaderElement, chartLegendElement) {
        var svgContainerHeightCalc = this.height - chartHeaderElement - chartLegendElement - Padding * 2;
        this.svgHeight = svgContainerHeightCalc <= 0 ? 0 : svgContainerHeightCalc;
        this.centerX = chartWrapperElement / 2;
    };
    PoChartDynamicTypeComponent.prototype.calculateTotalValue = function () {
        this.totalValue =
            this.type === PoChartType.Gauge
                ? 100
                : this.series.reduce(function (previousValue, serie) { return previousValue + serie.value; }, 0);
    };
    __decorate([
        ViewChild('chartBody', { static: true }),
        __metadata("design:type", ElementRef)
    ], PoChartDynamicTypeComponent.prototype, "chartBody", void 0);
    __decorate([
        ViewChild('svgContainer', { static: true }),
        __metadata("design:type", ElementRef)
    ], PoChartDynamicTypeComponent.prototype, "svgContainer", void 0);
    PoChartDynamicTypeComponent = __decorate([
        Directive()
    ], PoChartDynamicTypeComponent);
    return PoChartDynamicTypeComponent;
}());
export { PoChartDynamicTypeComponent };
