// @po-ui/ng-components/esm5/lib/components/po-loading/po-loading-overlay/po-loading-overlay-base.component.js
import { __decorate, __metadata } from "tslib";
import { Input, Directive } from '@angular/core';
import { convertToBoolean } from './../../../utils/util';
import { PoLanguageService } from '../../../services/po-language/po-language.service';
export var poLoadingOverlayLiteralsDefault = {
    en: {
        loading: 'Loading'
    },
    es: {
        loading: 'Cargando'
    },
    pt: {
        loading: 'Carregando'
    },
    ru: {
        loading: 'погрузка'
    }
};
/**
 *
 * @description
 *
 * Este componente mostra ao usuário uma imagem de _loading_ e bloqueia a página inteira ou o container escolhido,
 * enquanto aguarda a resposta de alguma requisição.
 */
var PoLoadingOverlayBaseComponent = /** @class */ (function () {
    function PoLoadingOverlayBaseComponent(languageService) {
        this.languageService = languageService;
        this._screenLock = false;
        this.text = this.getTextDefault();
    }
    Object.defineProperty(PoLoadingOverlayBaseComponent.prototype, "screenLock", {
        get: function () {
            return this._screenLock;
        },
        /**
         * @optional
         *
         * @description
         *
         * Define se o *overlay* será aplicado a um *container* ou a página inteira.
         *
         * Para utilizar o componente como um *container*, o elemento pai deverá receber uma posição relativa, por exemplo:
         *
         * ```
         * <div style="position: relative">
         *
         *  <po-chart [p-series]="[{ value: 10, category: 'Example' }]">
         *  </po-chart>
         *
         *  <po-loading-overlay>
         *  </po-loading-overlay>
         * </div>
         * ```
         *
         * @default `false`
         */
        set: function (screenLock) {
            this._screenLock = convertToBoolean(screenLock);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoLoadingOverlayBaseComponent.prototype, "text", {
        get: function () {
            return this._text;
        },
        /**
         * @optional
         *
         * @description
         *
         * Texto a ser exibido no componente.
         *
         * > O valor padrão será traduzido acordo com o idioma configurado no [**PoI18n**](/documentation/po-i18n) ou navegador.
         *
         * @default `Carregando`
         */
        set: function (value) {
            this._text = value || this.getTextDefault();
        },
        enumerable: true,
        configurable: true
    });
    PoLoadingOverlayBaseComponent.prototype.getTextDefault = function () {
        var language = this.languageService.getShortLanguage();
        return poLoadingOverlayLiteralsDefault[language].loading;
    };
    PoLoadingOverlayBaseComponent.ctorParameters = function () { return [
        { type: PoLanguageService }
    ]; };
    __decorate([
        Input('p-screen-lock'),
        __metadata("design:type", Boolean),
        __metadata("design:paramtypes", [Boolean])
    ], PoLoadingOverlayBaseComponent.prototype, "screenLock", null);
    __decorate([
        Input('p-text'),
        __metadata("design:type", String),
        __metadata("design:paramtypes", [String])
    ], PoLoadingOverlayBaseComponent.prototype, "text", null);
    PoLoadingOverlayBaseComponent = __decorate([
        Directive(),
        __metadata("design:paramtypes", [PoLanguageService])
    ], PoLoadingOverlayBaseComponent);
    return PoLoadingOverlayBaseComponent;
}());
export { PoLoadingOverlayBaseComponent };
