// @po-ui/ng-components/esm2015/lib/components/po-field/po-clean/po-clean-base.component.js
import { __decorate, __metadata } from "tslib";
import { ElementRef, EventEmitter, Input, Output, Directive } from '@angular/core';
/**
 * @docsPrivate
 *
 * Este componente é de uso interno utilizado por componentes de entrada de dados com o objetivo de resetar as informações do model.
 *
 * Por padrão limpa o valor do campo e executa o método onChangePropagate, caso tenha a necessidade de tratar a função de limpar o campo,
 * deve implementar a interface PoClean.
 */
let PoCleanBaseComponent = class PoCleanBaseComponent {
    constructor() {
        /** Valor que será atribuído ao campo quando for clicado no po-clean. */
        this.defaultValue = '';
        /**
         * Evento disparado quando executada ação do po-clean.
         * Este evento deve ser usado para avisar para o componente que está usando o po-clean, que o botão foi disparado,
         * e provavelmente será preciso emitir o evento para atualizar o model.
         */
        this.changeEvent = new EventEmitter();
    }
    clear() {
        this.setInputValue(this.defaultValue);
        this.changeEvent.emit(this.defaultValue);
    }
    showIcon() {
        return (this.defaultValue !== this.getInputValue() && this.hasCleanAttr() && !this.isDisabled() && !this.isReadonly());
    }
    // Este método verifica se o componente pai possui a propriedade clean diferente de vazio,
    // ou seja, se o po-clean deve ser usado.
    hasCleanAttr() {
        return this.parentComponent.clean;
    }
    // Este método verifica se o componente pai está desabilitado.
    isDisabled() {
        return this.parentComponent.disabled;
    }
    // Este método verifica se o componente pai está somente leitura.
    isReadonly() {
        return this.parentComponent.readonly;
    }
};
__decorate([
    Input('p-element-ref'),
    __metadata("design:type", ElementRef)
], PoCleanBaseComponent.prototype, "inputRef", void 0);
__decorate([
    Input('p-default-value'),
    __metadata("design:type", String)
], PoCleanBaseComponent.prototype, "defaultValue", void 0);
__decorate([
    Output('p-change-event'),
    __metadata("design:type", EventEmitter)
], PoCleanBaseComponent.prototype, "changeEvent", void 0);
PoCleanBaseComponent = __decorate([
    Directive()
], PoCleanBaseComponent);
export { PoCleanBaseComponent };
