// @po-ui/ng-components/esm5/lib/components/po-tree-view/po-tree-view.component.js
import { __decorate, __extends, __metadata } from "tslib";
import { ChangeDetectionStrategy, Component } from '@angular/core';
import { PoTreeViewBaseComponent } from './po-tree-view-base.component';
import { PoTreeViewService } from './services/po-tree-view.service';
/**
 * @docsExtends PoTreeViewBaseComponent
 *
 * @example
 *
 * <example name="po-tree-view-basic" title="PO Tree View Basic">
 *  <file name="sample-po-tree-view-basic/sample-po-tree-view-basic.component.html"> </file>
 *  <file name="sample-po-tree-view-basic/sample-po-tree-view-basic.component.ts"> </file>
 * </example>
 *
 * <example name="po-tree-view-labs" title="PO Tree View Labs">
 *  <file name="sample-po-tree-view-labs/sample-po-tree-view-labs.component.html"> </file>
 *  <file name="sample-po-tree-view-labs/sample-po-tree-view-labs.component.ts"> </file>
 * </example>
 *
 * <example name="po-tree-view-folder-structure" title="PO Tree View - Folder Structure">
 *  <file name="sample-po-tree-view-folder-structure/sample-po-tree-view-folder-structure.component.html"> </file>
 *  <file name="sample-po-tree-view-folder-structure/sample-po-tree-view-folder-structure.component.ts"> </file>
 * </example>
 *
 * <example name="po-tree-view-supermarket" title="PO Tree View - Supermarket">
 *  <file name="sample-po-tree-view-supermarket/sample-po-tree-view-supermarket.component.html"> </file>
 *  <file name="sample-po-tree-view-supermarket/sample-po-tree-view-supermarket.component.ts"> </file>
 * </example>
 */
var PoTreeViewComponent = /** @class */ (function (_super) {
    __extends(PoTreeViewComponent, _super);
    function PoTreeViewComponent(treeViewService) {
        var _this = _super.call(this) || this;
        _this.treeViewService = treeViewService;
        return _this;
    }
    Object.defineProperty(PoTreeViewComponent.prototype, "hasItems", {
        get: function () {
            return !!(this.items && this.items.length);
        },
        enumerable: true,
        configurable: true
    });
    PoTreeViewComponent.prototype.ngOnInit = function () {
        var _this = this;
        this.treeViewService.onExpand().subscribe(function (treeViewItem) {
            _this.emitExpanded(treeViewItem);
        });
        this.treeViewService.onSelect().subscribe(function (treeViewItem) {
            _this.emitSelected(treeViewItem);
        });
    };
    PoTreeViewComponent.prototype.trackByFunction = function (index) {
        return index;
    };
    PoTreeViewComponent.ctorParameters = function () { return [
        { type: PoTreeViewService }
    ]; };
    PoTreeViewComponent = __decorate([
        Component({
            selector: 'po-tree-view',
            template: "<po-container *ngIf=\"hasItems\" p-no-padding p-no-shadow>\n  <ul class=\"po-tree-view\">\n    <po-tree-view-item *ngFor=\"let item of items; trackBy: trackByFunction\" [p-item]=\"item\" [p-selectable]=\"selectable\">\n    </po-tree-view-item>\n  </ul>\n</po-container>\n",
            changeDetection: ChangeDetectionStrategy.OnPush,
            providers: [PoTreeViewService]
        }),
        __metadata("design:paramtypes", [PoTreeViewService])
    ], PoTreeViewComponent);
    return PoTreeViewComponent;
}(PoTreeViewBaseComponent));
export { PoTreeViewComponent };
