// @po-ui/ng-components/esm5/lib/components/po-dynamic/po-dynamic-form/po-dynamic-form-validation/po-dynamic-form-validation.service.js
import { __assign, __decorate, __extends, __metadata, __read, __spread } from "tslib";
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { map } from 'rxjs/operators';
import { PoDynamicFormOperation } from '../po-dynamic-form-operation/po-dynamic-form-operation';
var PoDynamicFormValidationService = /** @class */ (function (_super) {
    __extends(PoDynamicFormValidationService, _super);
    function PoDynamicFormValidationService(http) {
        return _super.call(this, http) || this;
    }
    PoDynamicFormValidationService.prototype.sendFieldChange = function (field, value) {
        var _this = this;
        var changedValue = { property: field.property, value: value };
        return this.execute(field.validate, changedValue).pipe(map(function (validateFields) { return _this.setFieldDefaultIfEmpty(validateFields); }));
    };
    PoDynamicFormValidationService.prototype.sendFormChange = function (validate, field, value) {
        var _this = this;
        var changedValue = { property: field.property, value: value };
        return this.execute(validate, changedValue).pipe(map(function (validateFields) { return _this.setFormDefaultIfEmpty(validateFields); }));
    };
    PoDynamicFormValidationService.prototype.updateFieldsForm = function (validatedFields, fields) {
        if (validatedFields === void 0) { validatedFields = []; }
        if (fields === void 0) { fields = []; }
        return __spread(validatedFields).reduce(function (updatedFields, validatedField) {
            var index = updatedFields.findIndex(function (field) { return field.property === validatedField.property; });
            var hasProperty = index >= 0;
            if (hasProperty) {
                updatedFields[index] = __assign(__assign({}, fields[index]), validatedField);
            }
            return updatedFields;
        }, __spread(fields));
    };
    PoDynamicFormValidationService.prototype.setFieldDefaultIfEmpty = function (validateFields) {
        return (validateFields || {
            field: {}
        });
    };
    PoDynamicFormValidationService.ctorParameters = function () { return [
        { type: HttpClient }
    ]; };
    PoDynamicFormValidationService = __decorate([
        Injectable(),
        __metadata("design:paramtypes", [HttpClient])
    ], PoDynamicFormValidationService);
    return PoDynamicFormValidationService;
}(PoDynamicFormOperation));
export { PoDynamicFormValidationService };
