// @po-ui/ng-components/esm5/lib/components/po-grid/po-grid-head/po-grid-head.component.js
import { __decorate, __metadata } from "tslib";
import { Component, Input, ViewChild, ElementRef, ChangeDetectorRef } from '@angular/core';
var PoGridHeadComponent = /** @class */ (function () {
    function PoGridHeadComponent(changeDetectorRef) {
        this.changeDetectorRef = changeDetectorRef;
        this.align = 'left';
        this.boolean = false;
        this.position = '';
        this.tabIndex = -1;
    }
    Object.defineProperty(PoGridHeadComponent.prototype, "width", {
        get: function () {
            return this._width ? this._width + "px" : '100%';
        },
        set: function (value) {
            this._width = value;
        },
        enumerable: true,
        configurable: true
    });
    PoGridHeadComponent.ctorParameters = function () { return [
        { type: ChangeDetectorRef }
    ]; };
    __decorate([
        Input('p-align'),
        __metadata("design:type", String)
    ], PoGridHeadComponent.prototype, "align", void 0);
    __decorate([
        Input('p-freeze'),
        __metadata("design:type", Boolean)
    ], PoGridHeadComponent.prototype, "boolean", void 0);
    __decorate([
        Input('p-position'),
        __metadata("design:type", String)
    ], PoGridHeadComponent.prototype, "position", void 0);
    __decorate([
        Input('p-tab-index'),
        __metadata("design:type", Number)
    ], PoGridHeadComponent.prototype, "tabIndex", void 0);
    __decorate([
        Input('p-width'),
        __metadata("design:type", Object),
        __metadata("design:paramtypes", [Object])
    ], PoGridHeadComponent.prototype, "width", null);
    __decorate([
        Input('p-title'),
        __metadata("design:type", String)
    ], PoGridHeadComponent.prototype, "title", void 0);
    PoGridHeadComponent = __decorate([
        Component({
            selector: 'po-grid-head',
            template: "<div>\n  <div [id]=\"position\" class=\"po-grid-head-content po-grid-head-{{ align }}\" tabindex=\"{{ title ? 0 : -1 }}\">\n    <span>{{ title || '&nbsp;' }}</span>\n  </div>\n</div>\n"
        }),
        __metadata("design:paramtypes", [ChangeDetectorRef])
    ], PoGridHeadComponent);
    return PoGridHeadComponent;
}());
export { PoGridHeadComponent };
